--- a/net/minecraft/world/entity/animal/allay/AllayAi.java
+++ b/net/minecraft/world/entity/animal/allay/AllayAi.java
@@ -112,6 +_,16 @@
         Optional<GlobalPos> likedNoteblockPos = brain.getMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
         if (likedNoteblockPos.isPresent()) {
             GlobalPos position = likedNoteblockPos.get();
+            // Luminol start - Do not like item if they were out of current tickregion
+            final Level targetLevel = allay.level().getServer().getLevel(position.dimension());
+            final BlockPos targetPos = position.pos();
+
+            // thread checks
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(targetLevel, targetPos)) {
+                brain.eraseMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION); // The memory value is not being belong to current tick region anymore
+                return Optional.empty();
+            }
+            // Luminol end
             if (shouldDepositItemsAtLikedNoteblock(allay, brain, position)) {
                 return Optional.of(new BlockPosTracker(position.pos().above()));
             }
