--- a/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
@@ -55,7 +_,7 @@
                     || belowState.is(Blocks.DIORITE)) {
                     BlockState softState = !currentState.isAir() && !isLiquid(currentState) ? currentState : Blocks.SAND.defaultBlockState();
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) {
                         BlockPos relativePos = pos.relative(direction);
                         BlockState relativeState = level.getBlockState(relativePos);
                         if (relativeState.isAir() || isLiquid(relativeState)) {
