--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
@@ -49,7 +_,7 @@
             }
         }
 
-        ObjectArrayList<BlockPos> list = new ObjectArrayList<>(set.stream().toList());
+        ObjectArrayList<BlockPos> list = new ObjectArrayList<>(set); // Leaf - paw optimization - TODO: use array
         RandomSource randomSource = RandomSource.create(level.getSeed()).forkPositional().at(pieces.calculateBoundingBox().getCenter());
         Util.shuffle(list, randomSource);
         int min = Math.min(set.size(), randomSource.nextInt(5, 8));
