--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
@@ -258,7 +_,7 @@
                     for (int y = 0; y < 3; y++) {
                         int pos = getRoomIndex(x, y, z);
                         if (roomGrid[pos] != null) {
-                            for (Direction direction : Direction.values()) {
+                            for (Direction direction : Direction.VALUES) {
                                 int neighX = x + direction.getStepX();
                                 int neighY = y + direction.getStepY();
                                 int neighZ = z + direction.getStepZ();
