--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -191,6 +_,18 @@
         return GameEventListenerRegistry.NOOP;
     }
 
+    private boolean slimeChunk;
+    private boolean hasComputedSlimeChunk;
+
+    public boolean isSlimeChunk() {
+        if (!hasComputedSlimeChunk) {
+            hasComputedSlimeChunk = true;
+            slimeChunk = net.edenor.foldenor.secureseed.WorldgenCryptoRandom.seedSlimeChunk(chunkPos.x, chunkPos.z).nextInt(10) == 0;
+        }
+
+        return slimeChunk;
+    }
+
     public abstract BlockState getBlockState(final int x, final int y, final int z); // Paper
 
     @Nullable
