--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -17,6 +_,7 @@
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 import java.util.regex.Pattern;
+import net.edenor.foldenor.secureseed.Globals;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.RegistryAccess;
@@ -137,7 +_,17 @@
         String levelSeed = this.get("level-seed", "");
         boolean generateStructures = this.get("generate-structures", true);
         long seed = WorldOptions.parseSeed(levelSeed).orElse(WorldOptions.randomSeed());
-        this.worldOptions = new WorldOptions(seed, generateStructures, false);
+        // Foldenor start - Feature secure seed
+        String featureSeedString = this.get("feature-level-seed", "");
+        long[] featureSeed;
+        if (featureSeedString.isEmpty()) {
+            featureSeed = Globals.createRandomWorldSeed();
+        } else {
+            featureSeed = Globals.parseSeed(featureSeedString).orElseGet(Globals::createRandomWorldSeed);
+        }
+
+        this.worldOptions = new WorldOptions(seed, featureSeed, generateStructures, false);
+        // Foldenor end - Feature secure seed
         this.worldDimensionData = new DedicatedServerProperties.WorldDimensionData(
             this.get("generator-settings", s -> GsonHelper.parse(!s.isEmpty() ? s : "{}"), new JsonObject()),
             this.get("level-type", v -> v.toLowerCase(Locale.ROOT), WorldPresets.NORMAL.identifier().toString())
