--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -19,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.chat.Component;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
@@ -149,7 +_,19 @@
     }
 
     private static int teleportToEntity(final CommandSourceStack source, final Collection<? extends Entity> entities, final Entity destination) throws CommandSyntaxException {
+        if (source.getPlayer() != null){
+            ServerPlayer sender = source.getPlayer();
+            if (sender.isSpectator() && sender.getCamera() != sender){
+                sender.setCamera(sender);
+            }
+        }
         for (Entity entity : entities) {
+            if (entity.isSpectator()){
+                ServerPlayer sender = (ServerPlayer) entity;
+                if (sender.getCamera() != sender){
+                    sender.setCamera(sender);
+                }
+            }
             io.papermc.paper.threadedregions.TeleportUtils.teleport(entity, false, destination, Float.valueOf(destination.getYRot()), Float.valueOf(destination.getXRot()), Entity.TELEPORT_FLAG_LOAD_CHUNK, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND, null); // Folia - region threading
         }
 
@@ -175,11 +_,23 @@
         final @Nullable Coordinates rotation,
         final @Nullable LookAt lookAt
     ) throws CommandSyntaxException {
+        if (source.getPlayer() != null){
+            ServerPlayer sender = source.getPlayer();
+            if (sender.isSpectator() && sender.getCamera() != sender){
+                sender.setCamera(sender);
+            }
+        }
         Vec3 pos = destination.getPosition(source);
         Vec2 rot = rotation == null ? null : rotation.getRotation(source);
 
         for (Entity entity : entities) {
             Set<Relative> relatives = getRelatives(destination, rotation, entity.level().dimension() == level.dimension());
+            if (entity.isSpectator()){
+                ServerPlayer sender = (ServerPlayer) entity;
+                if (sender.isSpectator() && sender.getCamera() != sender){
+                    sender.setCamera(sender);
+                }
+            }
             if (rot == null) {
                 performTeleport(source, entity, level, pos.x, pos.y, pos.z, relatives, entity.getYRot(), entity.getXRot(), lookAt);
             } else {
@@ -250,6 +_,9 @@
                 Float yawFinal = Float.valueOf(newYRot);
                 Float pitchFinal = Float.valueOf(newXRot);
                 victim.getBukkitEntity().taskScheduler.schedule((Entity nmsEntity) -> {
+                    if (nmsEntity instanceof ServerPlayer player && player.isSpectator() && player.getCamera() != player){
+                        player.setCamera(player);
+                    }
                     nmsEntity.stopRiding();
                     nmsEntity.teleportAsync(
                             level, posFinal, yawFinal, pitchFinal, Vec3.ZERO,
