--- a/net/minecraft/world/entity/monster/Slime.java
+++ b/net/minecraft/world/entity/monster/Slime.java
@@ -320,8 +_,8 @@
             }
 
             ChunkPos chunkPos = ChunkPos.containing(pos);
-            boolean slimeChunk = level.getMinecraftWorld().paperConfig().entities.spawning.allChunksAreSlimeChunks || WorldgenRandom.seedSlimeChunk(chunkPos.x(), chunkPos.z(), ((WorldGenLevel) level).getSeed(), level.getMinecraftWorld().spigotConfig.slimeSeed).nextInt(10) == 0; // Paper
-                // Paper start - Replace rules for Height in Slime Chunks
+            boolean slimeChunk = level.getMinecraftWorld().paperConfig().entities.spawning.allChunksAreSlimeChunks || level.getChunk(chunkPos.x(), chunkPos.z()).isSlimeChunk(); // Spigot // Paper
+            // Paper start - Replace rules for Height in Slime Chunks
                 final double maxHeightSlimeChunk = level.getMinecraftWorld().paperConfig().entities.spawning.slimeSpawnHeight.slimeChunk.maximum;
                 if (random.nextInt(10) == 0 && slimeChunk && pos.getY() < maxHeightSlimeChunk) {
                 // Paper end - Replace rules for Height in Slime Chunks
