--- a/net/minecraft/network/Utf8String.java
+++ b/net/minecraft/network/Utf8String.java
@@ -33,24 +_,19 @@
     }
 
     public static void write(final ByteBuf output, final CharSequence value, final int maxLength) {
+        // Paper start - optimize string encoding (from krypton)
+        // Mojang _almost_ gets it right, but stumbles at the finish line...
         if (value.length() > maxLength) {
             throw new EncoderException("String too big (was " + value.length() + " characters, max " + maxLength + ")");
+        }
+        int utf8Bytes = ByteBufUtil.utf8Bytes(value);
+        int maxBytesPermitted = ByteBufUtil.utf8MaxBytes(value);
+        if (utf8Bytes > maxBytesPermitted) {
+            throw new EncoderException("String too big (was " + utf8Bytes + " bytes encoded, max " + maxBytesPermitted + ")");
         } else {
-            int maxEncodedValueLength = ByteBufUtil.utf8MaxBytes(value);
-            ByteBuf tmp = output.alloc().buffer(maxEncodedValueLength);
-
-            try {
-                int bytesWritten = ByteBufUtil.writeUtf8(tmp, value);
-                int maxAllowedEncodedLength = ByteBufUtil.utf8MaxBytes(maxLength);
-                if (bytesWritten > maxAllowedEncodedLength) {
-                    throw new EncoderException("String too big (was " + bytesWritten + " bytes encoded, max " + maxAllowedEncodedLength + ")");
-                }
-
-                VarInt.write(output, bytesWritten);
-                output.writeBytes(tmp);
-            } finally {
-                tmp.release();
-            }
+            VarInt.write(output, utf8Bytes);
+            output.writeCharSequence(value, StandardCharsets.UTF_8);
         }
+        // Paper end
     }
 }
