--- a/net/minecraft/network/VarInt.java
+++ b/net/minecraft/network/VarInt.java
@@ -51,6 +_,41 @@
     }
 
     public static ByteBuf write(final ByteBuf output, int value) {
+        // Gale start - Velocity - optimized VarInt#write
+        if ((value & 0xFFFFFF80) == 0) {
+            output.writeByte(value);
+        } else if ((value & 0xFFFFC000) == 0) {
+            int w = (value & 0x7F) << 8
+                    | (value >>> 7)
+                    | 0x00008000;
+            output.writeShort(w);
+        } else if ((value & 0xFFE00000) == 0) {
+            int w = (value & 0x7F) << 16
+                    | (value & 0x3F80) << 1
+                    | (value >>> 14)
+                    | 0x00808000;
+            output.writeMedium(w);
+        } else if ((value & 0xF0000000) == 0) {
+            int w = (value & 0x7F) << 24
+                    | ((value & 0x3F80) << 9)
+                    | (value & 0x1FC000) >> 6
+                    | (value >>> 21)
+                    | 0x80808000;
+            output.writeInt(w);
+        } else {
+            int w = (value & 0x7F) << 24
+                    | (value & 0x3F80) << 9
+                    | (value & 0x1FC000) >> 6
+                    | ((value >>> 21) & 0x7F)
+                    | 0x80808080;
+            output.writeInt(w);
+            output.writeByte(value >>> 28);
+        }
+        return output;
+    }
+
+    static ByteBuf writeOld(ByteBuf output, int value) { // public -> package-private
+        // Gale end - Velocity - optimized VarInt#write
         // Paper start - Optimize VarInts
         // Peel the one and two byte count cases explicitly as they are the most common VarInt sizes
         // that the proxy will write, to improve inlining.
