--- a/io/papermc/paper/threadedregions/TeleportUtils.java
+++ b/io/papermc/paper/threadedregions/TeleportUtils.java
@@ -1,6 +_,7 @@
 package io.papermc.paper.threadedregions;
 
 import ca.spottedleaf.concurrentutil.completable.CallbackCompletable;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec3;
 import org.bukkit.Location;
@@ -25,6 +_,15 @@
                 }
                 final boolean scheduled = from.getBukkitEntity().taskScheduler.schedule(
                     (final Entity realFrom) -> {
+
+                        if (realFrom instanceof ServerPlayer player){
+                            if (player.isSpectator()){
+                                if (player.isSpectator() && player.getCamera() != player){
+                                    player.setCamera(player);
+                                }
+                            }
+                        }
+
                         final Vec3 pos = new Vec3(
                             loc.getX(), loc.getY(), loc.getZ()
                         );
