--- a/net/minecraft/world/level/levelgen/structure/placement/RandomSpreadStructurePlacement.java
+++ b/net/minecraft/world/level/levelgen/structure/placement/RandomSpreadStructurePlacement.java
@@ -67,8 +_,9 @@
     public ChunkPos getPotentialStructureChunk(long seed, int regionX, int regionZ) {
         int i = Math.floorDiv(regionX, this.spacing);
         int i1 = Math.floorDiv(regionZ, this.spacing);
-        WorldgenRandom worldgenRandom = new WorldgenRandom(new LegacyRandomSource(0L));
-        worldgenRandom.setLargeFeatureWithSalt(seed, i, i1, this.salt());
+        WorldgenRandom worldgenRandom = new net.edenor.foldenor.secureseed.WorldgenCryptoRandom(
+                i, i1, net.edenor.foldenor.secureseed.Globals.Salt.POTENTIONAL_FEATURE, this.salt
+        );
         int i2 = this.spacing - this.separation;
         int i3 = this.spreadType.evaluate(worldgenRandom, i2);
         int i4 = this.spreadType.evaluate(worldgenRandom, i2);
