--- a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
+++ b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
@@ -40,13 +_,22 @@
 
             @Override
             public void encode(final B output, final CustomPacketPayload value) {
+                // Leaves start - protocol core
+                if (value instanceof org.leavesmc.leaves.protocol.core.LeavesCustomPayload payload) {
+                    org.leavesmc.leaves.protocol.core.LeavesProtocolManager.encode(output, payload);
+                    return;
+                }
+                // Leaves end - protocol core
                 this.writeCap(output, value.type(), value);
             }
 
             @Override
             public CustomPacketPayload decode(final B input) {
                 Identifier identifier = input.readIdentifier();
-                return (CustomPacketPayload)this.findCodec(identifier).decode(input);
+                // Leaves start - protocol core
+                var payload = org.leavesmc.leaves.protocol.core.LeavesProtocolManager.decode(identifier, input);
+                return java.util.Objects.requireNonNullElseGet(payload, () -> this.findCodec(identifier).decode(input));
+                // Leaves end - protocol core
             }
         };
     }
