--- a/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
+++ b/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
@@ -9,7 +_,7 @@
         return CustomPacketPayload.codec((payload, buf) -> {
             // Paper start
             // Always write data
-            buf.writeBytes(payload.data);
+            if (payload.data != null) buf.writeBytes(payload.data); // Leaves - nullable
         }, buf -> {
             int length = buf.readableBytes();
             if (length >= 0 && length <= maxPayloadSize) {
