--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -27,7 +_,7 @@
     private final String namespace;
     private final String path;
 
-    private ResourceLocation(String namespace, String path) {
+    public ResourceLocation(String namespace, String path) { // Leaves - private -> public
         assert isValidNamespace(namespace);
 
         assert isValidPath(path);
