--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -126,6 +_,7 @@
     private Raid.RaidStatus status;
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos = Optional.empty();
+    private boolean isBarDirty; // DivineMC - lithium: ai.raid
 
     // Folia start - make raids thread-safe
     public boolean ownsRaid(ServerLevel world) {
@@ -286,6 +_,12 @@
 
     public void tick(ServerLevel level) {
         if (!this.isStopped()) {
+            // DivineMC start - lithium: ai.raid
+            if (this.isBarDirty) {
+                this.updateBossbarInternal();
+                this.isBarDirty = false;
+            }
+            // DivineMC end
             if (this.status == Raid.RaidStatus.ONGOING) {
                 boolean flag = this.active;
                 this.active = level.hasChunkAt(this.center);
@@ -603,9 +_,15 @@
         }
     }
 
+    // DivineMC start - lithium: ai.raid
     public void updateBossbar() {
+        this.isBarDirty = true;
+    }
+
+    private void updateBossbarInternal() {
         this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
     }
+    // DivineMC end
 
     public float getHealthOfLivingRaiders() {
         float f = 0.0F;
