--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -126,7 +_,7 @@
     }
 
     private boolean getNeighborSignal(final SignalGetter level, final BlockPos pos, final Direction pushDirection) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) {
             if (direction != pushDirection && level.hasSignal(pos.relative(direction), direction)) {
                 return true;
             }
@@ -137,7 +_,7 @@
         } else {
             BlockPos above = pos.above();
 
-            for (Direction directionx : Direction.values()) {
+            for (Direction directionx : Direction.VALUES) {
                 if (directionx != Direction.DOWN && level.hasSignal(above.relative(directionx), directionx)) {
                     return true;
                 }
