--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -92,7 +_,7 @@
     // Paper start - Allow ArmorStands not to tick
     public boolean canTick = true;
     public boolean canTickSetByAPI = false;
-    private boolean noTickEquipmentDirty = false;
+    public boolean noTickEquipmentDirty = false;
     // Paper end - Allow ArmorStands not to tick
 
     public ArmorStand(final EntityType<? extends ArmorStand> type, final Level level) {
@@ -458,7 +_,7 @@
         this.playBrokenSound();
         // this.dropAllDeathLoot(level, source); // CraftBukkit - moved down
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             ItemStack itemStack = this.equipment.get(slot); // Paper - move equipment removal past event call
             if (!itemStack.isEmpty() && !EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
                 this.drops.add(new DefaultDrop(itemStack, stack -> Block.popResource(this.level(), this.blockPosition().above(), stack))); // CraftBukkit - add to drops // Paper - Restore vanilla drops behavior; mirror so we can destroy it later - though this call site was safe & spawn drops correctly}
@@ -467,7 +_,7 @@
         // Paper start - move equipment removal past event call
         org.bukkit.event.entity.EntityDeathEvent event = this.dropAllDeathLoot(level, source);
         if (!event.isCancelled()) {
-            for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+            for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
                 this.equipment.set(slot, ItemStack.EMPTY);
             }
         }
@@ -525,8 +_,9 @@
     public void tick() {
         if (!this.canTick) {
             if (this.noTickEquipmentDirty) {
-                this.noTickEquipmentDirty = false;
+                //this.noTickEquipmentDirty = false; // Leaf - Lithium - equipment tracking - move down
                 this.detectEquipmentUpdates();
+                this.noTickEquipmentDirty = false;  // Leaf - Lithium - equipment tracking
             }
 
             return;
