--- a/net/minecraft/world/entity/ai/sensing/FrogAttackablesSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/FrogAttackablesSensor.java
@@ -15,7 +_,11 @@
 
     @Override
     protected boolean isMatchingEntity(final ServerLevel level, final LivingEntity body, final LivingEntity mob) {
-        return Sensor.isEntityAttackable(level, body, mob) && Frog.canEat(mob) && !this.isUnreachableAttackTarget(body, mob) && mob.closerThan(body, 10.0);
+        return Frog.canEat(mob)
+                && !body.getBrain().hasMemoryValue(MemoryModuleType.HAS_HUNTING_COOLDOWN)
+                && mob.closerThan(body, 10.0)
+                && Sensor.isEntityAttackable(level, body, mob)
+                && !this.isUnreachableAttackTarget(body, mob); // Sepals - Rearrange the attackable conditions;
     }
 
     private boolean isUnreachableAttackTarget(final LivingEntity body, final LivingEntity mob) {
