--- a/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
@@ -26,17 +_,39 @@
 
     @Override
     protected void doTick(final ServerLevel level, final LivingEntity body) {
-        List<Player> players = level.getLocalPlayers() // Folia - region threading
-            .stream()
-            .filter(EntitySelector.NO_SPECTATORS)
-            .filter(player -> body.closerThan(player, this.getFollowDistance(body)))
-            .sorted(Comparator.comparingDouble(body::distanceToSqr))
-            .collect(Collectors.toList());
+        // Leaf start - Remove stream in PlayerSensor
+        List<Player> players = new java.util.ArrayList<>();
+        for (Player serverPlayer : level.players()) {
+            if (!EntitySelector.NO_SPECTATORS.test(serverPlayer)) {
+                continue;
+            }
+            if (!body.closerThan(serverPlayer, this.getFollowDistance(body))) {
+                continue;
+            }
+
+            players.add(serverPlayer);
+        }
+        players.sort(Comparator.comparingDouble(body::distanceToSqr));
+        // Leaf end - Remove stream in PlayerSensor
         Brain<?> brain = body.getBrain();
         brain.setMemory(MemoryModuleType.NEAREST_PLAYERS, players);
-        List<Player> visiblePlayers = players.stream().filter(livingEntity -> isEntityTargetable(level, body, livingEntity)).collect(Collectors.toList());
+        // Leaf start - Remove stream in PlayerSensor
+        List<Player> visiblePlayers = new java.util.ArrayList<>(players.size());
+        for (Player player : players) {
+            if (isEntityTargetable(level, body, player)) {
+                visiblePlayers.add(player);
+            }
+        }
+        // Leaf end - Remove stream in PlayerSensor
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_PLAYER, visiblePlayers.isEmpty() ? null : visiblePlayers.get(0));
-        List<Player> visibleAttackablePlayers = visiblePlayers.stream().filter(livingEntity -> isEntityAttackable(level, body, livingEntity)).toList();
+        // Leaf start - Remove stream in PlayerSensor
+        List<Player> visibleAttackablePlayers = new java.util.ArrayList<>(visiblePlayers.size());
+        for (Player player : visiblePlayers) {
+            if (isEntityAttackable(level, body, player)) {
+                visibleAttackablePlayers.add(player);
+            }
+        }
+        // Leaf end - Remove stream in PlayerSensor
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYERS, visibleAttackablePlayers);
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYER, visibleAttackablePlayers.isEmpty() ? null : visibleAttackablePlayers.get(0));
     }
