--- a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -4,6 +_,8 @@
 import com.google.common.collect.Lists;
 import java.util.List;
 import java.util.Set;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.GlobalPos;
 import net.minecraft.resources.ResourceKey;
@@ -22,6 +_,15 @@
 
     @Override
     protected void doTick(final ServerLevel level, final Villager body) {
+        // Leaves start - skip secondary POI sensor if absent
+        if (FoldenorConfig.skipSecondaryPOISensorIfAbsent) {
+            var secondaryPoi = body.getVillagerData().profession().value().secondaryPoi();
+            if (secondaryPoi.isEmpty()) {
+                body.getBrain().eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
+                return;
+            }
+        }
+        // Leaves end - skip secondary POI sensor if absent
         ResourceKey<Level> dimensionType = level.dimension();
         BlockPos center = body.blockPosition();
         List<GlobalPos> jobSites = Lists.newArrayList();
