--- a/com/mojang/datafixers/types/Type.java
+++ b/com/mojang/datafixers/types/Type.java
@@ -93,10 +_,7 @@
     }
 
     public TypeTemplate template() {
-        if (template == null) {
-            template = buildTemplate();
-        }
-        return template;
+        return buildTemplate(); // Paper - Saturn: Avoid unnecessary object creation
     }
 
     public abstract TypeTemplate buildTemplate();
@@ -114,10 +_,7 @@
     }
 
     public final Codec<A> codec() {
-        if (codec == null) {
-            codec = buildCodec();
-        }
-        return codec;
+        return buildCodec(); // Paper - Saturn: Avoid unnecessary object creation
     }
 
     protected abstract Codec<A> buildCodec();
