--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,9 +_,24 @@
 
     public boolean areEnoughDeepSleeping(final int sleepPercentageNeeded, final List<ServerPlayer> players) {
         // CraftBukkit start
-        int deepSleepers = (int)players.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping).count();
-        boolean anyDeepSleep = players.stream().anyMatch(Player::isSleepingLongEnough);
-        return anyDeepSleep && deepSleepers >= this.sleepersNeeded(sleepPercentageNeeded);
+        // Leaf start - Remove stream and double iteration in enough deep sleeping player check
+        int count = 0;
+        boolean anyPlayerSleeping = false;
+
+        for (ServerPlayer player : players) {
+            final boolean isSleepingLongEnough = player.isSleepingLongEnough();
+
+            if (isSleepingLongEnough) {
+                anyPlayerSleeping = true;
+            }
+
+            if (isSleepingLongEnough || player.fauxSleeping) {
+                count++;
+            }
+        }
+
+        return anyPlayerSleeping && count >= this.sleepersNeeded(sleepPercentageNeeded);
+        // Leaf end - Remove stream and double iteration in enough deep sleeping player check
         // CraftBukkit end
     }
 
