--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -225,9 +_,10 @@
     );
     private static final boolean DEFAULT_SEEN_CREDITS = false;
     private static final boolean DEFAULT_SPAWN_EXTRA_PARTICLES_ON_FALL = false;
+    public final it.unimi.dsi.fastutil.objects.Object2ObjectMap<net.minecraft.world.waypoints.WaypointTransmitter, net.minecraft.world.waypoints.WaypointTransmitter.Connection> canvas$activeWaypoints = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // Canvas - region threading
     public ServerGamePacketListenerImpl connection;
     private final MinecraftServer server;
-    public final ServerPlayerGameMode gameMode;
+    public ServerPlayerGameMode gameMode;
     private final PlayerAdvancements advancements;
     private final ServerStatsCounter stats;
     private float lastRecordedHealthAndAbsorption = Float.MIN_VALUE;
@@ -438,6 +_,8 @@
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder viewDistanceHolder = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder();
 
+    public net.minecraft.network.Connection internalConnection; // Leaves - protocol core
+
     @Override
     public final boolean moonrise$isRealPlayer() {
         return this.isRealPlayer;
@@ -1067,9 +_,10 @@
             }
             // CraftBukkit end
         } catch (Throwable var4) {
-            CrashReport crashReport = CrashReport.forThrowable(var4, "Ticking player");
+           CrashReport crashReport = CrashReport.forThrowable(var4, "Ticking player");
             CrashReportCategory crashReportCategory = crashReport.addCategory("Player being ticked");
             this.fillCrashReportCategory(crashReportCategory);
+            MinecraftServer.LOGGER.error(var4.getMessage(), var4); // Foldenor - more info
             throw new ReportedException(crashReport);
         }
     }
@@ -1381,7 +_,7 @@
                     this.getInventory().getNonEquipmentItems().set(i, net.minecraft.world.item.ItemStack.EMPTY);
                 }
             }
-            for (final EquipmentSlot value : EquipmentSlot.VALUES) {
+            for (final EquipmentSlot value : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
                 if (this.getInventory().equipment.has(value) && !shouldKeepDeathEventItem(event, this.getInventory().equipment.get(value))) {
                     this.getInventory().equipment.set(value, net.minecraft.world.item.ItemStack.EMPTY);
                 }
