--- /dev/null
+++ b/net/edenor/foldenor/config/FoldenorConfig.java
@@ -1,0 +_,282 @@
+package net.edenor.foldenor.config;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.triassic.linearpaper.region.RegionFileFormat;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.entity.EntityType;
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+public class FoldenorConfig {
+    protected static final String HEADER = "This is the main configuration file for Foldenor.";
+    protected static File CONFIG_FILE;
+    public static YamlConfiguration config;
+
+    public static int version;
+    static boolean verbose;
+
+    public static boolean sendNullEntityPackets = false;
+
+    public static boolean appleskinProtocol = false;
+
+    public static boolean useAlternateKeepAlive = false;
+
+    public static int piglinSpawnChancePersentInPortal = 100;
+
+    public static boolean useVirtualThreadForAsyncScheduler = false;
+
+    public static boolean skipMapItemUpdatesIfNoBukkitRender = true;
+
+    public static int entityActivationCheckFrequency = 20;
+
+    public static int linearFlushFrequency = 10;
+    public static int linearFlushThreads = 1;
+    public static RegionFileFormat regionFormat = RegionFileFormat.ANVIL;
+    public static int linearCompressionLevel = 1;
+
+    public static boolean dearEnabled;
+    public static int startDistance;
+    public static int startDistanceSquared;
+    public static int maximumActivationPrio;
+    public static int activationDistanceMod;
+
+    public static boolean skipAIForNonAwareMob = true;
+
+    public static boolean throttleHopperWhenFullEnabled = false;
+    public static int throttleHopperWhenFullSkipTicks = 0;
+
+    public static boolean asyncPlayerDataSaveEnabled = false;
+
+    public static void init(File configFile) {
+        init(configFile, true);
+    }
+
+    public static void reload(File configFile) {
+        init(configFile, false);
+    }
+
+    private static void init(File configFile, boolean setup) {
+        CONFIG_FILE = configFile;
+        config = new YamlConfiguration();
+        if (configFile.exists()) {
+            try {
+                config.load(CONFIG_FILE);
+            } catch (InvalidConfigurationException ex){
+                Bukkit.getLogger().log(Level.SEVERE, "Could not load foldenor.yml, please correct your syntax errors", ex);
+                throw Throwables.propagate(ex);
+            } catch (IOException ignore) {}
+        }
+        config.options().header(HEADER);
+        config.options().copyDefaults(true);
+        verbose = getBoolean("verbose", false);
+
+        version = getInt("config-version", 1);
+        set("config-version", 1);
+
+        readConfig();
+    }
+
+    protected static void log(String s) {
+        if (verbose) {
+            log(Level.INFO, s);
+        }
+    }
+
+    protected static void log(Level level, String s) {
+        Bukkit.getLogger().log(level, s);
+    }
+
+    static void readConfig() {
+        readNetworkSettings();
+
+        readOptimizationSettings();
+
+        readMiscSettings();
+
+        readLinearRegion();
+
+        try {
+            dynamicActivationOfBrains();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        try {
+            config.save(CONFIG_FILE);
+        } catch (IOException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not save " + CONFIG_FILE, ex);
+        }
+    }
+
+    private static void readLinearRegion() {
+        linearFlushFrequency = getInt("region-format.linear.flush-frequency", linearFlushFrequency);
+        linearFlushThreads = getInt("region-format.linear.flush-max-threads", linearFlushThreads);
+
+        if (linearFlushThreads < 0)
+            linearFlushThreads = Math.max(Runtime.getRuntime().availableProcessors() + linearFlushThreads, 1);
+        else
+            linearFlushThreads = Math.max(linearFlushThreads, 1);
+
+        regionFormat = RegionFileFormat.fromString(getString("region-format.type", regionFormat.name()));
+        if (regionFormat.equals(RegionFileFormat.INVALID)) {
+            log(Level.SEVERE, "Unknown region format in linear.yml: " + regionFormat);
+            log(Level.SEVERE, "Falling back to ANVIL region file format.");
+            regionFormat = RegionFileFormat.ANVIL;
+        }
+
+        linearCompressionLevel = getInt("region-format.linear.compression-level", linearCompressionLevel);
+        if (linearCompressionLevel > 23 || linearCompressionLevel < 1) {
+            log(Level.SEVERE, "Linear region compression level should be between 1 and 22 in linear.yml: " + linearCompressionLevel);
+            log(Level.SEVERE, "Falling back to compression level 1.");
+            linearCompressionLevel = 1;
+        }
+    }
+
+    protected static void readNetworkSettings(){
+        sendNullEntityPackets = getBoolean("network.send-null-entity-packets", sendNullEntityPackets);
+        appleskinProtocol = getBoolean("network.appleskin-protocol", appleskinProtocol);
+        useAlternateKeepAlive = getBoolean("network.use-alternate-keepalive", useAlternateKeepAlive);
+    }
+
+    private static void readOptimizationSettings() {
+        piglinSpawnChancePersentInPortal = getInt("optimizations.piglin-spawn-chance-persent-in-portal", 100,
+            "Reduces piglin spawn in portal, by reducing change to spawn");
+        skipMapItemUpdatesIfNoBukkitRender = getBoolean("optimizations.skip_map_item_updates_if_no_bukkit_render", skipMapItemUpdatesIfNoBukkitRender);
+        entityActivationCheckFrequency = getInt("optimizations.entity-activation-check-frequency", 20);
+        skipAIForNonAwareMob = getBoolean("optimizations.skip-ai-for-non-aware-mob", skipAIForNonAwareMob);
+        throttleHopperWhenFullEnabled = getBoolean("optimizations.throttle-hopper-when-full.enabled", throttleHopperWhenFullEnabled);
+        throttleHopperWhenFullSkipTicks = getInt("optimizations.throttle-hopper-when-full.skip-ticks", throttleHopperWhenFullSkipTicks);
+    }
+
+    private static void readMiscSettings() {
+        useVirtualThreadForAsyncScheduler = getBoolean("optimizations.use-virtual-thread-for-async-scheduler", useVirtualThreadForAsyncScheduler,
+            "Use the new Virtual Thread introduced in JDK 21 for CraftAsyncScheduler.");
+        asyncPlayerDataSaveEnabled = getBoolean("misc.async-playerdata-save.enabled", asyncPlayerDataSaveEnabled);
+    }
+
+    private static void dynamicActivationOfBrains() throws IOException {
+        dearEnabled = getBoolean("dab.enabled", true);
+        startDistance = getInt("dab.start-distance", 12,
+                "This value determines how far away an entity has to be",
+                "from the player to start being effected by DEAR.");
+        startDistanceSquared = startDistance * startDistance;
+        maximumActivationPrio = getInt("dab.max-tick-freq", 20,
+                "This value defines how often in ticks, the furthest entity",
+                "will get their pathfinders and behaviors ticked. 20 = 1s");
+        activationDistanceMod = getInt("dab.activation-dist-mod", 8,
+                "This value defines how much distance modifies an entity's",
+                "tick frequency. freq = (distanceToPlayer^2) / (2^value)",
+                "If you want further away entities to tick less often, use 7.",
+                "If you want further away entities to tick more often, try 9.");
+
+        for (EntityType<?> entityType : BuiltInRegistries.ENTITY_TYPE) {
+            entityType.dabEnabled = true; // reset all, before setting the ones to true
+        }
+        getStringList("dab.blacklisted-entities", Collections.emptyList(), "A list of entities to ignore for activation")
+                .forEach(name -> EntityType.byString(name).ifPresentOrElse(entityType -> {
+                    entityType.dabEnabled = false;
+                    }, () -> MinecraftServer.LOGGER.warn("Unknown entity \"" + name + "\"")));
+    }
+
+    protected static void set(String path, Object val) {
+        config.addDefault(path, val);
+        config.set(path, val);
+    }
+
+    protected static String getString(String path, String def, String... comment) {
+        config.addDefault(path, def);
+        return config.getString(path, config.getString(path));
+    }
+
+    protected static boolean getBoolean(String path, boolean def, String... comment) {
+        config.addDefault(path, def);
+        return config.getBoolean(path, config.getBoolean(path));
+    }
+
+    protected static double getDouble(String path, double def, String... comment) {
+        config.addDefault(path, def);
+        return config.getDouble(path, config.getDouble(path));
+    }
+
+    protected static int getInt(String path, int def, String... comment) {
+        config.addDefault(path, def);
+        return config.getInt(path, config.getInt(path));
+    }
+
+    protected static <T> List getList(String path, T def, String... comment) {
+        config.addDefault(path, def);
+        return config.getList(path, config.getList(path));
+    }
+
+    protected static List<String> getStringList(String path, List<String> def, String... comment) {
+        config.addDefault(path, def);
+        return config.getStringList(path);
+    }
+
+    static Map<String, Object> getMap(String path, Map<String, Object> def, String... comment) {
+        if (def != null && config.getConfigurationSection(path) == null) {
+            config.addDefault(path, def);
+            return def;
+        }
+        return toMap(config.getConfigurationSection(path));
+    }
+
+    protected static String getString(String path, String def) {
+        config.addDefault(path, def);
+        return config.getString(path, config.getString(path));
+    }
+
+    protected static boolean getBoolean(String path, boolean def) {
+        config.addDefault(path, def);
+        return config.getBoolean(path, config.getBoolean(path));
+    }
+
+    protected static double getDouble(String path, double def) {
+        config.addDefault(path, def);
+        return config.getDouble(path, config.getDouble(path));
+    }
+
+    protected static int getInt(String path, int def) {
+        config.addDefault(path, def);
+        return config.getInt(path, config.getInt(path));
+    }
+
+    protected static <T> List getList(String path, T def) {
+        config.addDefault(path, def);
+        return config.getList(path, config.getList(path));
+    }
+
+    static Map<String, Object> getMap(String path, Map<String, Object> def) {
+        if (def != null && config.getConfigurationSection(path) == null) {
+            config.addDefault(path, def);
+            return def;
+        }
+        return toMap(config.getConfigurationSection(path));
+    }
+
+    protected static Map<String, Object> toMap(ConfigurationSection section) {
+        ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();
+        if (section != null) {
+            for (String key : section.getKeys(false)) {
+                Object obj = section.get(key);
+                if (obj != null) {
+                    builder.put(key, obj instanceof ConfigurationSection val ? toMap(val) : obj);
+                }
+            }
+        }
+        return builder.build();
+    }
+}
