--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -10,6 +_,8 @@
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.BiPredicate;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.SharedConstants;
 import net.minecraft.core.GlobalPos;
 import net.minecraft.core.Holder;
@@ -86,6 +_,7 @@
 import net.minecraft.world.item.trading.TradeSet;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
+import net.minecraft.world.level.portal.TeleportTransition;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
@@ -182,6 +_,8 @@
         (villager, poiType) -> poiType.is(PoiTypes.MEETING)
     );
 
+    public long nextGolemPanic = -1; // Pufferfish
+
     public Villager(final EntityType<? extends Villager> type, final Level level) {
         this(type, level, VillagerType.PLAINS);
     }
@@ -244,6 +_,21 @@
         return this.assignProfessionWhenSpawned;
     }
 
+    // Leaves start - fixes a memory leak when villagers get moved to another world
+    @Override
+    protected Entity teleportCrossDimension(ServerLevel oldLevel, ServerLevel newLevel, TeleportTransition teleportTransition) {
+        if (FoldenorConfig.villagersDontReleaseMemoryFix) {
+            this.releaseAllPois();
+            this.getBrain().eraseMemory(MemoryModuleType.HOME);
+            this.getBrain().eraseMemory(MemoryModuleType.JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.POTENTIAL_JOB_SITE);
+            this.getBrain().eraseMemory(MemoryModuleType.MEETING_POINT);
+            this.refreshBrain(teleportTransition.newLevel());
+        }
+        return super.teleportCrossDimension(oldLevel, newLevel, teleportTransition);
+    }
+    // Leaves end - fixes a memory leak when villagers get moved to another world
+
     // Paper start - EAR 2
     @Override
     public void inactiveTick() {
@@ -263,6 +_,8 @@
     }
     // Paper end - EAR 2
 
+    private int behaviorTick = 0; // Pufferfish
+
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         // Paper start - EAR 2
@@ -272,7 +_,11 @@
         // Paper end - EAR 2
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
-        if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
+        // Pufferfish start
+        if (!inactive && this.behaviorTick++ % this.activatedPriority == 0) {
+            this.getBrain().tick((ServerLevel) this.level(), this); // Paper
+        }
+        // Pufferfish end
         profiler.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
