--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1018,10 +_,10 @@
         if (flush) {
             for (ServerLevel level : this.getAllLevels()) {
                 String storageName = level.getChunkSource().chunkMap.getStorageName();
-                LOGGER.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName));
+                LOGGER.info("ThreadedChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName)); // LinearPaper
             }
 
-            LOGGER.info("ThreadedAnvilChunkStorage: All dimensions are saved");
+            LOGGER.info("ThreadedChunkStorage: All dimensions are saved"); // LinearPaper
         }
 
         return result;
@@ -1986,6 +_,7 @@
         }
 
         profiler.popPush("server gui refresh");
+        org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handleTick(region.getCurrentTick()); // Leaves - protocol
 
         if (false) for (Runnable tickable : this.tickables) { // Folia - region threading - TODO WTF is this?
             tickable.run();
