--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -993,10 +_,10 @@
         // CraftBukkit end
         if (flush) {
             for (ServerLevel serverLevel : this.getAllLevels()) {
-                LOGGER.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", serverLevel.getChunkSource().chunkMap.getStorageName());
+                LOGGER.info("ThreadedChunkStorage ({}): All chunks are saved", serverLevel.getChunkSource().chunkMap.getStorageName()); // LinearPaper
             }
 
-            LOGGER.info("ThreadedAnvilChunkStorage: All dimensions are saved");
+            LOGGER.info("ThreadedChunkStorage: All dimensions are saved"); // LinearPaper
         }
 
         return flag;
@@ -1950,6 +_,7 @@
         }
 
         profilerFiller.popPush("server gui refresh");
+        org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handleTick(region.getCurrentTick()); // Leaves - protocol
 
         if (false) for (Runnable runnable : this.tickables) { // Folia - region threading - TODO WTF is this?
             runnable.run();
