--- a/net/minecraft/world/phys/shapes/MinecartCollisionContext.java
+++ b/net/minecraft/world/phys/shapes/MinecartCollisionContext.java
@@ -6,6 +_,7 @@
 import net.minecraft.world.level.block.BaseRailBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.RailShape;
+import org.bukkit.Bukkit;
 import org.jspecify.annotations.Nullable;
 
 public class MinecartCollisionContext extends EntityCollisionContext {
@@ -18,22 +_,24 @@
     }
 
     private void setupContext(AbstractMinecart minecart) {
-        BlockPos currentBlockPosOrRailBelow = minecart.getCurrentBlockPosOrRailBelow();
-        BlockState blockState = minecart.level().getBlockState(currentBlockPosOrRailBelow);
-        boolean isRail = BaseRailBlock.isRail(blockState);
-        if (isRail) {
-            this.ingoreBelow = currentBlockPosOrRailBelow.below();
-            RailShape railShape = blockState.getValue(((BaseRailBlock)blockState.getBlock()).getShapeProperty());
-            if (railShape.isSlope()) {
-                this.slopeIgnore = switch (railShape) {
-                    case ASCENDING_EAST -> currentBlockPosOrRailBelow.east();
-                    case ASCENDING_WEST -> currentBlockPosOrRailBelow.west();
-                    case ASCENDING_NORTH -> currentBlockPosOrRailBelow.north();
-                    case ASCENDING_SOUTH -> currentBlockPosOrRailBelow.south();
-                    default -> null;
-                };
+        Bukkit.getRegionScheduler().execute(null, minecart.getBukkitEntity().getLocation(), () -> {
+            BlockPos currentBlockPosOrRailBelow = minecart.getCurrentBlockPosOrRailBelow();
+            BlockState blockState = minecart.level().getBlockState(currentBlockPosOrRailBelow);
+            boolean isRail = BaseRailBlock.isRail(blockState);
+            if (isRail) {
+                this.ingoreBelow = currentBlockPosOrRailBelow.below();
+                RailShape railShape = blockState.getValue(((BaseRailBlock)blockState.getBlock()).getShapeProperty());
+                if (railShape.isSlope()) {
+                    this.slopeIgnore = switch (railShape) {
+                        case ASCENDING_EAST -> currentBlockPosOrRailBelow.east();
+                        case ASCENDING_WEST -> currentBlockPosOrRailBelow.west();
+                        case ASCENDING_NORTH -> currentBlockPosOrRailBelow.north();
+                        case ASCENDING_SOUTH -> currentBlockPosOrRailBelow.south();
+                        default -> null;
+                    };
+                }
             }
-        }
+        });
     }
 
     @Override
