--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -341,7 +_,7 @@
     }
 
     private boolean isEquipped(final Item item) {
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             ItemStack itemStack = this.getItemBySlot(slot);
             Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
             if (itemStack.is(item) && equippable != null && equippable.slot() == slot) {
