--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -81,6 +_,11 @@
     public static void setWalkAndLookTargetMemories(
         final LivingEntity walker, final PositionTracker target, final float speedModifier, final int closeEnoughDistance
     ) {
+        // Luminol - Do not set walk target if target position is out of current tick region
+        if (!target.checkThread(walker.level())) {
+            return;
+        }
+        // Luminol end
         WalkTarget walkTarget = new WalkTarget(target, speedModifier, closeEnoughDistance);
         walker.getBrain().setMemory(MemoryModuleType.LOOK_TARGET, target);
         walker.getBrain().setMemory(MemoryModuleType.WALK_TARGET, walkTarget);
@@ -114,10 +_,33 @@
 
     public static SectionPos findSectionClosestToVillage(final ServerLevel level, final SectionPos center, final int radius) {
         int distToVillage = level.sectionsToVillage(center);
-        return SectionPos.cube(center, radius)
-            .filter(s -> level.sectionsToVillage(s) < distToVillage)
-            .min(Comparator.comparingInt(level::sectionsToVillage))
-            .orElse(center);
+        // Leaf start - Remove stream in BehaviorUtils
+        SectionPos closestSection = center;
+        int closestDistance = radius;
+
+        final int lowerX = center.getX() - radius;
+        final int lowerY = center.getY() - radius;
+        final int lowerZ = center.getZ() - radius;
+        final int upperX = center.getX() + radius;
+        final int upperY = center.getY() + radius;
+        final int upperZ = center.getZ() + radius;
+
+        for (int x = lowerX; x <= upperX; x++) {
+            for (int z = lowerZ; z <= upperZ; z++) {
+                for (int y = lowerY; y <= upperY; y++) {
+                    SectionPos pos = SectionPos.of(x, y, z);
+                    int distance = level.sectionsToVillage(pos);
+
+                    if (distance < closestDistance) {
+                        closestDistance = distance;
+                        closestSection = pos;
+                    }
+                }
+            }
+        }
+
+        return closestSection;
+        // Leaf end - Remove stream in BehaviorUtils
     }
 
     public static boolean isWithinAttackRange(final Mob body, final LivingEntity target, final int projectileAttackRangeMargin) {
