--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -1,5 +_,6 @@
 package io.papermc.paper.entity.activation;
 
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.entity.Entity;
@@ -36,6 +_,7 @@
 import net.minecraft.world.entity.vehicle.AbstractMinecart;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
 import org.spigotmc.SpigotWorldConfig;
 import java.util.List;
 import java.util.Set;
@@ -144,6 +_,8 @@
         maxRange = Math.min((world.spigotConfig.simulationDistance << 4) - 8, maxRange);
 
         for (final Player player : world.getLocalPlayers()) { // Folia - region threading
+            if (FoldenorConfig.entityActivationCheckFrequency > 1 && (player.getId() + io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick()) % FoldenorConfig
+                    .entityActivationCheckFrequency != 0) continue; // ShreddedPaper - Configurable entity activation check frequency
             player.activatedTick = io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick(); // Folia - region threading
             if (world.spigotConfig.ignoreSpectatorActivation && player.isSpectator()) {
                 continue;
@@ -174,6 +_,24 @@
                 }
 
                 ActivationRange.activateEntity(entity, bbByType); // Folia - threaded regions
+
+                // Pufferfish start
+                if (FoldenorConfig.dearEnabled && entity.getType().dabEnabled) {
+                    if (!entity.activatedPriorityReset) {
+                        entity.activatedPriorityReset = true;
+                        entity.activatedPriority = FoldenorConfig.maximumActivationPrio;
+                    }
+                    Vec3 playerVec = player.position();
+                    Vec3 entityVec = entity.position();
+                    double diffX = playerVec.x - entityVec.x, diffY = playerVec.y - entityVec.y, diffZ = playerVec.z - entityVec.z;
+                    int squaredDistance = (int) (diffX * diffX + diffY * diffY + diffZ * diffZ);
+                    entity.activatedPriority = squaredDistance > FoldenorConfig.startDistanceSquared ?
+                            Math.max(1, Math.min(squaredDistance >> FoldenorConfig.activationDistanceMod, entity.activatedPriority)) :
+                            1;
+                } else {
+                    entity.activatedPriority = 1;
+                }
+                // Pufferfish end
             }
         }
     }
@@ -184,13 +_,13 @@
      * @param entity
      */
     private static void activateEntity(final Entity entity, final AABB[] bbByType) { // Folia - threaded regions
-        if (io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() > entity.activatedTick) { // Folia - threaded regions
+        if (io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + Math.max(0, FoldenorConfig.entityActivationCheckFrequency) > entity.activatedTick ) {// Folia - threaded regions // ShreddedPaper - Configurable entity activation check frequency
             if (entity.defaultActivationState) {
-                entity.activatedTick = io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick(); // Folia - threaded regions
+                entity.activatedTick = io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + Math.max(0, FoldenorConfig.entityActivationCheckFrequency); // Folia - threaded regions // ShreddedPaper - Configurable entity activation check frequency
                 return;
             }
             if (bbByType[entity.activationType.ordinal()].intersects(entity.getBoundingBox())) { // Folia - threaded regions
-                entity.activatedTick = io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick(); // Folia - threaded regions
+                entity.activatedTick = io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + Math.max(0, FoldenorConfig.entityActivationCheckFrequency); // Folia - threaded regions // ShreddedPaper - Configurable entity activation check frequency
             }
         }
     }
@@ -231,7 +_,7 @@
         }
         // special cases.
         if (entity instanceof final LivingEntity living) {
-            if (living.onClimbable() || living.jumping || living.hurtTime > 0 || !living.activeEffects.isEmpty() || living.isFreezing()) {
+            if ( living.onClimableCached() || living.jumping || living.hurtTime > 0 || living.activeEffects.size() > 0) {// Paper
                 return 1;
             }
             if (entity instanceof final Mob mob && mob.getTarget() != null) {
