--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -23,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.storage.ChunkSystemRegionFile { // Paper - rewrite chunk system
+public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.storage.ChunkSystemRegionFile , abomination.IRegionFile{ // Paper - rewrite chunk system // Luminol - Configurable region file format
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int MAX_CHUNK_SIZE = 500 * 1024 * 1024; // Paper - don't write garbage data to disk if writing serialization fails
     private static final int SECTOR_BYTES = 4096;
@@ -131,7 +_,7 @@
         return this.recalculateCount.get();
     }
 
-    boolean recalculateHeader() throws IOException {
+    public boolean recalculateHeader() throws IOException { // Luminol - Configurable region file format // Luminol - Configurable region file format
         if (!this.canRecalcHeader) {
             return false;
         }
@@ -790,7 +_,7 @@
         }
     }
 
-    protected synchronized void write(final ChunkPos pos, final ByteBuffer data) throws IOException {
+    public synchronized void write(final ChunkPos pos, final ByteBuffer data) throws IOException {
         int offsetIndex = getOffsetIndex(pos);
         int offset = this.offsets.get(offsetIndex);
         int sectorNumber = getSectorNumber(offset);
@@ -908,7 +_,7 @@
         }
 
         @Override
-        public final void moonrise$write(final RegionFile regionFile) throws IOException {
+        public final void moonrise$write(final abomination.IRegionFile regionFile) throws IOException { // Luminol - Configurable region file format
             regionFile.write(this.pos, ByteBuffer.wrap(this.buf, 0, this.count));
         }
         // Paper end - rewrite chunk system
@@ -975,11 +_,11 @@
         return (x & 31) + (z & 31) * 32;
     }
 
-    synchronized boolean isOversized(int x, int z) {
+    public synchronized boolean isOversized(int x, int z) { // Luminol - Configurable region file format
         return this.oversized[getChunkIndex(x, z)] == 1;
     }
 
-    synchronized void setOversized(int x, int z, boolean oversized) throws IOException {
+    public synchronized void setOversized(int x, int z, boolean oversized) throws IOException { // Luminol - Configurable region file format
         final int offset = getChunkIndex(x, z);
         boolean previous = this.oversized[offset] == 1;
         this.oversized[offset] = (byte) (oversized ? 1 : 0);
@@ -1018,7 +_,7 @@
         return this.path.getParent().resolve(this.path.getFileName().toString().replaceAll("\\.mca$", "") + "_oversized_" + x + "_" + z + ".nbt");
     }
 
-    synchronized net.minecraft.nbt.CompoundTag getOversizedData(int x, int z) throws IOException {
+    public synchronized net.minecraft.nbt.CompoundTag getOversizedData(int x, int z) throws IOException { // Luminol - Configurable region file format
         Path file = getOversizedFile(x, z);
         try (DataInputStream out = new DataInputStream(new java.io.BufferedInputStream(new java.util.zip.InflaterInputStream(Files.newInputStream(file))))) {
             return net.minecraft.nbt.NbtIo.read((java.io.DataInput) out);
