--- a/net/minecraft/world/level/chunk/storage/RegionStorageInfo.java
+++ b/net/minecraft/world/level/chunk/storage/RegionStorageInfo.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.level.chunk.storage;
 
+import me.earthme.luminol.enums.EnumRegionFormat;
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.Level;
 
@@ -12,4 +_,14 @@
     public RegionStorageInfo withTypeSuffix(final String suffix) {
         return new RegionStorageInfo(this.level, this.dimension, this.type + suffix);
     }
+
+    // LinearPaper start
+    public EnumRegionFormat regionFormat() {
+        return FoldenorConfig.regionFormat;
+    }
+
+    public int linearCompressionLevel() {
+        return FoldenorConfig.linearCompressionLevel;
+    }
+    // LinearPaper end
 }
