--- a/net/minecraft/world/level/chunk/storage/SectionStorage.java
+++ b/net/minecraft/world/level/chunk/storage/SectionStorage.java
@@ -52,7 +_,7 @@
     private final Function<Runnable, R> factory;
     private final RegistryAccess registryAccess;
     private final ChunkIOErrorReporter errorReporter;
-    protected final LevelHeightAccessor levelHeightAccessor;
+    public final LevelHeightAccessor levelHeightAccessor;
     private final LongSet loadedChunks = new LongOpenHashSet();
     private final Long2ObjectMap<CompletableFuture<Optional<SectionStorage.PackedChunk<P>>>> pendingLoads = new Long2ObjectOpenHashMap<>();
     private final Object loadLock = new Object();
