--- a/folia-api/build.gradle.kts
+++ b/folia-api/build.gradle.kts
@@ -14,8 +_,8 @@
 val annotationsVersion = "26.0.2"
 val adventureVersion = "4.25.0"
 val bungeeCordChatVersion = "1.21-R0.2-deprecated+build.21"
-val slf4jVersion = "2.0.16"
-val log4jVersion = "2.24.1"
+val slf4jVersion = "2.0.17"
+val log4jVersion = "2.24.3"
 
 val apiAndDocs: Configuration by configurations.creating {
     attributes {
@@ -42,9 +_,9 @@
 
 dependencies {
     // api dependencies are listed transitively to API consumers
-    api("com.google.guava:guava:33.3.1-jre")
-    api("com.google.code.gson:gson:2.11.0")
-    api("org.yaml:snakeyaml:2.2")
+    api("com.google.guava:guava:33.4.0-jre")
+    api("com.google.code.gson:gson:2.12.1")
+    api("org.yaml:snakeyaml:2.3")
     api("org.joml:joml:1.10.8") {
         isTransitive = false // https://github.com/JOML-CI/JOML/issues/352
     }
@@ -66,9 +_,9 @@
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
 
-    api("org.apache.maven:maven-resolver-provider:3.9.6") // make API dependency for Paper Plugins
-    implementation("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
-    implementation("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
+    api("org.apache.maven:maven-resolver-provider:3.9.9") // make API dependency for Paper Plugins
+    compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.22")
+    compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.22")
 
     // Annotations - Slowly migrate to jspecify
     val annotations = "org.jetbrains:annotations:$annotationsVersion"
@@ -84,11 +_,15 @@
     // Test dependencies
     testImplementation("org.apache.commons:commons-lang3:3.17.0")
     testImplementation("org.junit.jupiter:junit-jupiter:5.12.2")
-    testImplementation("org.hamcrest:hamcrest:2.2")
-    testImplementation("org.mockito:mockito-core:5.14.1")
+    testImplementation("org.hamcrest:hamcrest:3.0")
+    testImplementation("org.mockito:mockito-core:5.16.1")
     testImplementation("org.ow2.asm:asm-tree:9.8")
-    mockitoAgent("org.mockito:mockito-core:5.14.1") { isTransitive = false } // configure mockito agent that is needed in newer java versions
+    mockitoAgent("org.mockito:mockito-core:5.16.1") { isTransitive = false } // configure mockito agent that is needed in newer java versions
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
+
+    // commons-lang3 is removed in maven-resolver-provider since 3.9.8
+    // Add this because bukkit api still need it.
+    compileOnly("org.apache.commons:commons-lang3:3.17.0")
 }
 
 val generatedDir: java.nio.file.Path = rootProject.layout.projectDirectory.dir("paper-api/src/generated/java").asFile.toPath()
@@ -102,17 +_,21 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../folia-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../folia-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../folia-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../folia-api/src/test/resources"))
         }
     }
 }
@@ -178,6 +_,14 @@
     apiVersion.set(rootProject.providers.gradleProperty("apiVersion"))
 }
 
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+
 tasks.jar {
     from(generateApiVersioningFile.flatMap { it.outputFile })
     manifest {
@@ -187,6 +_,11 @@
     }
 }
 
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+
 abstract class Services {
     @get:Inject
     abstract val fileSystemOperations: FileSystemOperations
@@ -199,8 +_,8 @@
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
-        "https://guava.dev/releases/33.3.1-jre/api/docs/",
-        "https://www.javadocs.dev/org.yaml/snakeyaml/2.2/",
+        "https://guava.dev/releases/33.4.0-jre/api/docs/",
+        "https://javadoc.io/doc/org.yaml/snakeyaml/2.3/",
         "https://www.javadocs.dev/org.jetbrains/annotations/$annotationsVersion/",
         "https://www.javadocs.dev/org.joml/joml/1.10.8/",
         "https://www.javadocs.dev/com.google.code.gson/gson/2.11.0",
@@ -214,7 +_,7 @@
         "https://jd.advntr.dev/text-logger-slf4j/$adventureVersion/",
         "https://www.javadocs.dev/org.slf4j/slf4j-api/$slf4jVersion/",
         "https://logging.apache.org/log4j/2.x/javadoc/log4j-api/",
-        "https://www.javadocs.dev/org.apache.maven.resolver/maven-resolver-api/1.7.3",
+        "https://javadoc.io/doc/org.apache.maven.resolver/maven-resolver-api/1.9.22",
     )
     options.tags("apiNote:a:API Note:")
 
@@ -238,6 +_,8 @@
             into("build/docs/javadoc")
         }
     }
+    options.addStringOption("Xdoclint:none", "-quiet")
+    options.addStringOption("-add-modules", "jdk.incubator.vector")
 }
 
 tasks.test {
