--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -4,6 +_,8 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Vec3i;
 import net.minecraft.server.level.ServerLevel;
@@ -99,6 +_,7 @@
     }
 
     public void recomputePath() {
+        if (!FoldenorConfig.mobRePathFinding) return;
         if (this.tick - this.timeLastRecompute <= 20L || !this.canUpdatePath()) { // Folia - region threading
             this.hasDelayedRecomputation = true;
         } else if (this.targetPos != null) {
@@ -447,6 +_,7 @@
     }
 
     public boolean shouldRecomputePath(final BlockPos pos) {
+        if (!FoldenorConfig.mobRePathFinding) return false;
         if (this.hasDelayedRecomputation) {
             return false;
         } else if (this.path != null && !this.path.isDone() && this.path.getNodeCount() != 0) {
