--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -434,7 +_,7 @@
 
         @Override
         public int[] getSlotsForFace(final Direction direction) {
-            return new int[0];
+            return me.titaniumtown.ArrayConstants.emptyIntArray; // Gale - JettPack - reduce array allocations
         }
 
         @Override
@@ -469,7 +_,7 @@
 
         @Override
         public int[] getSlotsForFace(final Direction direction) {
-            return direction == Direction.UP ? new int[]{0} : new int[0];
+            return direction == Direction.UP ? me.titaniumtown.ArrayConstants.zeroSingletonIntArray : me.titaniumtown.ArrayConstants.emptyIntArray; // Gale - JettPack - reduce array allocations
         }
 
         @Override
@@ -520,7 +_,7 @@
 
         @Override
         public int[] getSlotsForFace(final Direction direction) {
-            return direction == Direction.DOWN ? new int[]{0} : new int[0];
+            return direction == Direction.DOWN ? me.titaniumtown.ArrayConstants.zeroSingletonIntArray : me.titaniumtown.ArrayConstants.emptyIntArray; // Gale - JettPack - reduce array allocations
         }
 
         @Override
