--- a/net/minecraft/world/level/block/FrostedIceBlock.java
+++ b/net/minecraft/world/level/block/FrostedIceBlock.java
@@ -49,7 +_,7 @@
             if (brightness > 11 - state.getValue(AGE) - state.getLightDampening() && this.slightlyMelt(state, level, pos)) {
                 BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
 
-                for (Direction direction : Direction.values()) {
+                for (Direction direction : Direction.VALUES) {
                     neighborPos.setWithOffset(pos, direction);
                     BlockState neighbour = level.getBlockState(neighborPos);
                     if (neighbour.is(this) && !this.slightlyMelt(neighbour, level, neighborPos)) {
@@ -95,7 +_,7 @@
         int result = 0;
         BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) {
             neighborPos.setWithOffset(pos, direction);
             if (level.getBlockState(neighborPos).is(this)) {
                 if (++result >= limit) {
