--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -120,7 +_,7 @@
         int i = 7;
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) {
             mutableBlockPos.setWithOffset(pos, direction);
             i = Math.min(i, getDistanceAt(level.getBlockState(mutableBlockPos)) + 1);
             if (i == 1) {
