--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -4,6 +_,8 @@
 import com.mojang.serialization.MapCodec;
 import java.util.Map;
 import java.util.Optional;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleTypes;
@@ -69,7 +_,7 @@
         if (level.spigotConfig.enableZombiePigmenPortalSpawns && level.isSpawningMonsters() // Spigot
             && level.environmentAttributes().getValue(EnvironmentAttributes.NETHER_PORTAL_SPAWNS_PIGLINS, pos)
             && random.nextInt(2000) < level.getDifficulty().getId()
-            && level.anyPlayerCloseEnoughForSpawning(pos)) {
+            && level.anyPlayerCloseEnoughForSpawning(pos) && random.nextInt(100) <= FoldenorConfig.piglinSpawnChancePersentInPortal) {
             while (level.getBlockState(pos).is(this)) {
                 pos = pos.below();
             }
