--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -5,6 +_,8 @@
 import java.util.Map;
 import java.util.Optional;
 import javax.annotation.Nullable;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.BlockUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -65,10 +_,9 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.spigotConfig.enableZombiePigmenPortalSpawns && level.dimensionType().natural() // Spigot
-            && level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)
-            && random.nextInt(2000) < level.getDifficulty().getId()
-            && level.anyPlayerCloseEnoughForSpawning(pos)) {
+        if (level.spigotConfig.enableZombiePigmenPortalSpawns && level.dimensionType().natural()
+                && level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && random.nextInt(2000) < level.getDifficulty().getId()
+                && random.nextInt(100) <= FoldenorConfig.piglinSpawnChancePersentInPortal) { // Spigot // Foldenor
             while (level.getBlockState(pos).is(this)) {
                 pos = pos.below();
             }
