--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -109,11 +_,15 @@
         this.dirty.add(stat);
     }
 
+    // SparklyPaper start - Skip dirty stats copy when requesting player stats
+    /*
     private Set<Stat<?>> getDirty() {
         Set<Stat<?>> result = Sets.newHashSet(this.dirty);
         this.dirty.clear();
         return result;
     }
+    */
+    // SparklyPaper end
 
     public void parse(final DataFixer fixerUpper, final JsonElement element) {
         Dynamic<JsonElement> data = new Dynamic<>(JsonOps.INSTANCE, element);
@@ -140,9 +_,11 @@
     public void sendStats(final ServerPlayer player) {
         Object2IntMap<Stat<?>> statsToSend = new Object2IntOpenHashMap<>();
 
-        for (Stat<?> stat : this.getDirty()) {
+        for (Stat<?> stat : this.dirty) { // SparklyPaper - Skip dirty stats copy when requesting player stats
             statsToSend.put(stat, this.getValue(stat));
         }
+
+        this.dirty.clear(); // SparklyPaper - Skip dirty stats copy when requesting player stats
 
         player.connection.send(new ClientboundAwardStatsPacket(statsToSend));
     }
