--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -34,6 +_,8 @@
 
     @Override
     public void updateWaypoint(final WaypointTransmitter waypoint) {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         if (this.waypoints.contains(waypoint)) {
             Map<ServerPlayer, WaypointTransmitter.Connection> playerConnection = Tables.transpose(this.connections).row(waypoint);
@@ -51,6 +_,8 @@
 
     @Override
     public void untrackWaypoint(final WaypointTransmitter waypoint) {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         this.connections.column(waypoint).forEach((player, connection) -> connection.disconnect());
         Tables.transpose(this.connections).row(waypoint).clear();
         this.waypoints.remove(waypoint);
@@ -69,11 +_,15 @@
     }
 
     public void breakAllConnections() {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         this.connections.values().forEach(WaypointTransmitter.Connection::disconnect);
         this.connections.clear();
     }
 
     public void remakeConnections(final WaypointTransmitter waypoint) {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         if (!this.locatorBarEnabled) return; // Paper - optimize ServerWaypointManager with locator bar disabled
         for (ServerPlayer player : this.players) {
             this.createConnection(player, waypoint);
@@ -81,6 +_,8 @@
     }
 
     public Set<WaypointTransmitter> transmitters() {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         return this.waypoints;
     }
 
@@ -92,6 +_,8 @@
     }
 
     private void createConnection(final ServerPlayer player, final WaypointTransmitter waypoint) {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         if (player != waypoint) {
             if (isLocatorBarEnabledFor(player)) {
                 waypoint.makeWaypointConnectionWith(player).ifPresentOrElse(connection -> {
@@ -108,6 +_,8 @@
     }
 
     private void updateConnection(final ServerPlayer player, final WaypointTransmitter waypoint, final WaypointTransmitter.Connection connection) {
+        if (true)
+            throw new UnsupportedOperationException("Unsupported in region threading");
         if (player != waypoint) {
             if (isLocatorBarEnabledFor(player)) {
                 if (!connection.isBroken()) {
