--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -144,6 +_,12 @@
                         .ofType(String.class)
                         .defaultsTo("Unknown Server")
                         .describedAs("Name");
+
+                this.acceptsAll(asList("foldenor", "foldenor-settings"), "File for foldenor settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("foldenor.yml"))
+                        .describedAs("Yml file");
             }
         };
 
