--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
@@ -39,7 +_,7 @@
 
     @Override
     public CraftScoreboard getNewScoreboard() {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
+        //if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
         org.spigotmc.AsyncCatcher.catchOp("scoreboard creation"); // Spigot
         CraftScoreboard scoreboard = new CraftScoreboard(new ServerScoreboard(this.server));
         if (io.papermc.paper.configuration.GlobalConfiguration.get().scoreboards.trackPluginScoreboards) {
@@ -60,7 +_,7 @@
     }
 
     public void setPlayerBoard(CraftPlayer player, CraftScoreboard scoreboard) {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
+        //if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
         net.minecraft.world.scores.Scoreboard oldBoard = this.getPlayerBoard(player).getHandle();
         net.minecraft.world.scores.Scoreboard newBoard = scoreboard.getHandle();
         if (oldBoard == newBoard) {
