--- a/net/minecraft/core/BlockPos.java
+++ b/net/minecraft/core/BlockPos.java
@@ -333,7 +_,18 @@
         };
     }
 
+    // JettPack start - lithium: cached iterate outwards
+    private static final org.dreeam.leaf.util.cache.IterateOutwardsCache ITERATE_OUTWARDS_CACHE = new org.dreeam.leaf.util.cache.IterateOutwardsCache(50);
+    private static final it.unimi.dsi.fastutil.longs.LongList HOGLIN_PIGLIN_CACHE = ITERATE_OUTWARDS_CACHE.getOrCompute(8, 4, 8);
+    // JettPack end
+
     public static Iterable<BlockPos> withinManhattan(final BlockPos origin, final int reachX, final int reachY, final int reachZ) {
+        // JettPack start - lithium: cached iterate outwards
+        if (origin != org.dreeam.leaf.util.cache.IterateOutwardsCache.POS_ZERO) {
+            final it.unimi.dsi.fastutil.longs.LongList positions = reachX == 8 && reachY == 4 && reachZ == 8 ? HOGLIN_PIGLIN_CACHE : ITERATE_OUTWARDS_CACHE.getOrCompute(reachX, reachY, reachZ);
+            return new org.dreeam.leaf.util.cache.LongList2BlockPosMutableIterable(origin, positions);
+        }
+        // JettPack end
         int maxDepth = reachX + reachY + reachZ;
         int originX = origin.getX();
         int originY = origin.getY();
