--- a/net/minecraft/world/item/crafting/ShapedRecipePattern.java
+++ b/net/minecraft/world/item/crafting/ShapedRecipePattern.java
@@ -121,7 +_,7 @@
         }
 
         if (pattern.size() == bottom) {
-            return new String[0];
+            return me.titaniumtown.ArrayConstants.emptyStringArray; // Gale - JettPack - reduce array allocations
         } else {
             String[] result = new String[pattern.size() - bottom - top];
 
