--- a/net/minecraft/world/entity/ai/Brain.java
+++ b/net/minecraft/world/entity/ai/Brain.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.entity.ai;
 
+import com.github.cao.awa.catheter.Catheter;
+import com.github.cao.awa.sepals.entity.ai.task.composite.SepalsTaskStatus;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
