--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
@@ -2,6 +_,7 @@
 
 import java.util.function.Consumer;
 
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
 import org.bukkit.Bukkit;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.scheduler.BukkitTask;
@@ -29,6 +_,7 @@
     private final Plugin plugin;
     private final int id;
     private final long createdAt = System.nanoTime();
+    private ScheduledTask scheduledTask;
 
     CraftTask() {
         this(null, null, CraftTask.NO_REPEATING, CraftTask.NO_REPEATING);
@@ -64,7 +_,10 @@
 
     @Override
     public final Plugin getOwner() {
-        return this.plugin;
+        if (this.plugin == null && this.scheduledTask != null && this.scheduledTask.getOwningPlugin() != null)
+            return this.scheduledTask.getOwningPlugin();
+        else
+            return this.plugin;
     }
 
     @Override
@@ -115,12 +_,18 @@
 
     @Override
     public boolean isCancelled() {
-        return (this.period == CraftTask.CANCEL);
+        return scheduledTask != null ? scheduledTask.isCancelled() : (this.period == CraftTask.CANCEL);
     }
 
     @Override
     public void cancel() {
-        Bukkit.getScheduler().cancelTask(this.id);
+        //Bukkit.getScheduler().cancelTask(this.id);
+        if (this.scheduledTask != null)
+            scheduledTask.cancel();
+    }
+
+    protected void setScheduledTask(ScheduledTask scheduledTask) {
+        this.scheduledTask = scheduledTask;
     }
 
     /**
