--- a/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
@@ -97,7 +_,7 @@
         if (blockPathType == PathType.WATER) {
             BlockPos.MutableBlockPos reusablePos = new BlockPos.MutableBlockPos();
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) {
                 reusablePos.set(x, y, z).move(direction);
                 PathType pathType = context.getPathTypeFromState(reusablePos.getX(), reusablePos.getY(), reusablePos.getZ());
                 if (pathType == PathType.BLOCKED) {
