--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2976,6 +_,12 @@
     }
     // Paper end - Folia region threading API
 
+    // Leaves start - Photographer API
+    public static @NotNull org.leavesmc.leaves.entity.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Leaves end - Photographer API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
