--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -157,8 +_,9 @@
     public void handleConfigurationFinished(ServerboundFinishConfigurationPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.server);
         this.finishCurrentTask(JoinWorldTask.TYPE);
-        this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
+        // this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess()))); // Luminol - Async protocol switch - move down
 
+        Runnable afterSwitch = () -> { // Luminol - Async protocol switch
         try {
             PlayerList playerList = this.server.getPlayerList();
             if (playerList.getPlayer(this.gameProfile.getId()) != null) {
@@ -246,6 +_,18 @@
             this.connection.send(new ClientboundDisconnectPacket(DISCONNECT_REASON_INVALID_DATA));
             this.connection.disconnect(DISCONNECT_REASON_INVALID_DATA);
         }
+        // Luminol start - Async protocol switch
+        };
+        if (!net.edenor.foldenor.config.FoldenorConfig.asyncProtocolChange) {
+            this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
+            afterSwitch.run(); // directly run callback as we won't process any packet this time
+        } else {
+            this.connection.setupOutboundProtocolAsync(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())), () -> {
+                // push back
+                io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(afterSwitch);
+                }, false); // we will start auto read once we set up inbound handler at placeNewPlayer in PlayerList
+        }
+        // Luminol end
     }
 
     @Override
