--- a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.entity.ai.goal;
 
+import net.edenor.foldenor.config.FoldenorConfig;
+import net.edenor.foldenor.util.FoldenorUtils;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.core.particles.ItemParticleOption;
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.gamerules.GameRules;
@@ -34,6 +_,7 @@
 
     @Override
     public boolean canUse() {
+        if (FoldenorConfig.turtleEggSearch && blockToRemove == Blocks.TURTLE_EGG && !FoldenorUtils.rollPercentChance(FoldenorConfig.turtleEggSearchPercentChance)) return false;
         if (!getServerLevel(this.removerMob).getGameRules().get(GameRules.MOB_GRIEFING)) {
             return false;
         } else if (this.nextStartTick > 0) {
