--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -52,6 +_,24 @@
         return BlockPos.betweenClosedStream(area).map(this::getBlockState);
     }
 
+    // Leaf start - Optimize check nearby fire or lava on entity move
+    public default java.util.List<BlockState> getBlockStatesList(int minX, int minY, int minZ, int maxX, int maxY, int maxZ, java.util.List<BlockState> into) {
+        int rangeX = maxX - minX + 1;
+        int rangeY = maxY - minY + 1;
+        int rangeZ = maxZ - minZ + 1;
+
+        for (int x = 0; x < rangeX; x++) {
+            for (int y = 0; y < rangeY; y++) {
+                for (int z = 0; z < rangeZ; z++) {
+                    into.add(getBlockState(new BlockPos(minX + x, minY + y, minZ + z)));
+                }
+            }
+        }
+
+        return into;
+    }
+    // Leaf end - Optimize check nearby fire or lava on entity move
+
     default BlockHitResult isBlockInLine(ClipBlockStateContext context) {
         return traverseBlocks(
             context.getFrom(),
