--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level;
 
+import java.util.concurrent.CompletableFuture;
 import java.util.stream.Stream;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -9,6 +_,9 @@
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.SectionPos;
 import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.level.ChunkHolder;
+import net.minecraft.server.level.ChunkResult;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.world.attribute.EnvironmentAttributeReader;
@@ -17,6 +_,7 @@
 import net.minecraft.world.level.biome.BiomeManager;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.levelgen.Heightmap;
@@ -55,6 +_,12 @@
         return this.getBiomeManager().getBiome(pos);
     }
 
+    // Leaf start - cache getBiome
+    default Holder<Biome> getBiomeCached(net.minecraft.world.level.chunk.LevelChunk chunk, BlockPos pos) {
+        return this.getBiomeManager().getBiomeCached(chunk, pos);
+    }
+    // Leaf end - cache getBiome
+
     default Stream<BlockState> getBlockStatesIfLoaded(AABB aabb) {
         int floor = Mth.floor(aabb.minX);
         int floor1 = Mth.floor(aabb.maxX);
@@ -65,6 +_,18 @@
         return this.hasChunksAt(floor, floor2, floor4, floor1, floor3, floor5) ? this.getBlockStates(aabb) : Stream.empty();
     }
 
+    // Leaf start - Optimize check nearby fire or lava on entity move
+    default java.util.List<BlockState> getBlockStatesListIfLoaded(AABB box, java.util.List<BlockState> into) {
+        int i = Mth.floor(box.minX);
+        int j = Mth.floor(box.maxX);
+        int k = Mth.floor(box.minY);
+        int l = Mth.floor(box.maxY);
+        int m = Mth.floor(box.minZ);
+        int n = Mth.floor(box.maxZ);
+        return this.hasChunksAt(i, k, m, j, l, n) ? this.getBlockStatesList(i, k, m, j, l, n , into) : into;
+    }
+    // Leaf end - Optimize check nearby fire or lava on entity move
+
     @Override
     default int getBlockTint(BlockPos pos, ColorResolver colorResolver) {
         return colorResolver.getColor(this.getBiome(pos).value(), pos.getX(), pos.getZ());
@@ -72,9 +_,26 @@
 
     @Override
     default Holder<Biome> getNoiseBiome(int x, int y, int z) {
-        ChunkAccess chunk = this.getChunk(QuartPos.toSection(x), QuartPos.toSection(z), ChunkStatus.BIOMES, false);
+        ChunkAccess chunk = this.canvas$fasterChunkAccess(this, QuartPos.toSection(x), QuartPos.toSection(z), ChunkStatus.BIOMES, false); // Canvas
         return chunk != null ? chunk.getNoiseBiome(x, y, z) : this.getUncachedNoiseBiome(x, y, z);
     }
+
+    // Canvas start
+    private @Nullable ChunkAccess canvas$fasterChunkAccess(LevelReader instance, int x, int z, ChunkStatus chunkStatus, boolean create) {
+        if (!create && instance instanceof ServerLevel world) {
+            final ChunkHolder holder = (world.getChunkSource().chunkMap).getVisibleChunkIfPresent(ChunkPos.asLong(x, z));
+            if (holder != null) {
+                final CompletableFuture<ChunkResult<LevelChunk>> future = holder.getFullChunkFuture();
+                final ChunkResult<LevelChunk> either = future.getNow(null);
+                if (either != null) {
+                    final LevelChunk chunk = either.orElse(null);
+                    if (chunk != null) return chunk;
+                }
+            }
+        }
+        return instance.getChunk(x, z, chunkStatus, create);
+    }
+    // Canvas end
 
     Holder<Biome> getUncachedNoiseBiome(int x, int y, int z);
 
