--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -916,10 +_,7 @@
         for (int i = 0, len = entities.size(); i < len; ++i) {
             Entity entity = entities.get(i);
 
-            if (checkCanSee && source instanceof net.minecraft.server.level.ServerPlayer && entity instanceof net.minecraft.server.level.ServerPlayer
-                && !((net.minecraft.server.level.ServerPlayer) source).getBukkitEntity().canSee(((net.minecraft.server.level.ServerPlayer) entity).getBukkitEntity())) {
-                continue;
-            }
+            // Leaf - move up
 
             // !entity1.isRemoved() && entity1.blocksBuilding && (entity == null || !entity1.isPassengerOfSameVehicle(entity))
             // elide the last check since vanilla calls with entity = null
@@ -927,6 +_,12 @@
             if (entity.isRemoved() || !entity.blocksBuilding) {
                 continue;
             }
+            // Leaf end
+
+            if (checkCanSee && source instanceof net.minecraft.server.level.ServerPlayer && entity instanceof net.minecraft.server.level.ServerPlayer
+                    && !((net.minecraft.server.level.ServerPlayer) source).getBukkitEntity().canSee(((net.minecraft.server.level.ServerPlayer) entity).getBukkitEntity())) {
+                continue;
+            }
 
             if (net.minecraft.world.phys.shapes.Shapes.joinIsNotEmpty(collisionShape, net.minecraft.world.phys.shapes.Shapes.create(entity.getBoundingBox()), net.minecraft.world.phys.shapes.BooleanOp.AND)) {
                 return false;
@@ -1014,13 +_,17 @@
             }
         }
         // CraftBukkit end
-        if (this.isOutsideBuildHeight(pos)) {
+        // Leaf start - optimize get chunk
+        int x = pos.getX();
+        int y = pos.getY();
+        int z = pos.getZ();
+        if (this.isOutsideBuildHeight(y)) {
             return Blocks.VOID_AIR.defaultBlockState();
         } else {
-            ChunkAccess chunk = this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4);
-
-            return chunk == null ? null : chunk.getBlockState(pos);
+            ChunkAccess chunk = this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
+            return chunk.getBlockState(x, y, z);
         }
+        // Leaf end - optimize get chunk
     }
 
     @Override
@@ -1920,7 +_,7 @@
     public org.bukkit.entity.Entity[] getChunkEntities(int chunkX, int chunkZ) {
         ca.spottedleaf.moonrise.patches.chunk_system.level.entity.ChunkEntitySlices slices = ((ServerLevel)this).moonrise$getEntityLookup().getChunk(chunkX, chunkZ);
         if (slices == null) {
-            return new org.bukkit.entity.Entity[0];
+            return me.titaniumtown.ArrayConstants.emptyBukkitEntityArray; // Gale - JettPack - reduce array allocations
         }
 
         List<org.bukkit.entity.Entity> ret = new java.util.ArrayList<>();
@@ -1931,7 +_,7 @@
             }
         }
 
-        return ret.toArray(new org.bukkit.entity.Entity[0]);
+        return ret.toArray(me.titaniumtown.ArrayConstants.emptyBukkitEntityArray); // Gale - JettPack - reduce array allocations
     }
     // Paper end - rewrite chunk system
 
