--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -447,7 +_,7 @@
         final RandomSource random,
         final BlockPos pos
     ) {
-        Holder<Biome> biome = level.getBiome(pos);
+        Holder<Biome> biome = net.edenor.foldenor.config.FoldenorConfig.cacheBiomeMobSpawn ? level.getBiomeCached(null, pos) : level.getBiome(pos); // Leaf - cache getBiome
         return mobCategory == MobCategory.WATER_AMBIENT && biome.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && random.nextFloat() < 0.98F
             ? Optional.empty()
             : mobsAt(level, structureManager, generator, mobCategory, pos, biome).getRandom(random);
@@ -474,7 +_,7 @@
     ) {
         return isInNetherFortressBounds(pos, level, mobCategory, structureManager)
             ? NetherFortressStructure.FORTRESS_ENEMIES
-            : generator.getMobsAt(biome != null ? biome : level.getBiome(pos), structureManager, mobCategory, pos);
+                : generator.getMobsAt(biome != null ? biome : (net.edenor.foldenor.config.FoldenorConfig.cacheBiomeMobSpawn ? level.getBiomeCached(null, pos) : level.getBiome(pos)), structureManager, mobCategory, pos); // Leaf - cache getBiome
     }
 
     public static boolean isInNetherFortressBounds(
