--- a/folia-server/build.gradle.kts
+++ b/folia-server/build.gradle.kts
@@ -32,7 +_,22 @@
         }
     }
 
-    activeFork = fork
+    val foldenor = forks.register("foldenor") {
+        forks = fork
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = fork.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("foldenor-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchDir("foliaServer") {
+            upstreamPath = "folia-server"
+            excludes = setOf("src/minecraft", "paper-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("foldenor-server/folia-patches")
+            outputDir = rootDirectory.dir("folia-server")
+        }
+    }
+
+    activeFork = foldenor
 
 
     updatingMinecraft {
@@ -103,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java"); srcDir("../paper-server/src/generated/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../folia-server/src/main/java") }
+        resources { srcDir("../folia-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../folia-server/src/test/java") }
+        resources { srcDir("../folia-server/src/test/resources") }
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -134,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":folia-api"))
+    implementation(project(":foldenor-api"))
     implementation("ca.spottedleaf:concurrentutil:0.0.10")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -182,6 +_,19 @@
     // Spark
     implementation("me.lucko:spark-api:0.1-20240720.200737-2")
     implementation("me.lucko:spark-paper:1.10.152")
+
+    // LinearPaper start
+    implementation("com.github.luben:zstd-jni:1.5.6-6")
+    implementation("net.openhft:zero-allocation-hashing:0.16")
+    // LinearPaper end
+
+    implementation("com.github.ben-manes.caffeine:caffeine:3.1.8")
+    implementation("io.github.classgraph:classgraph:4.8.158")
+
+    //Idk why plugins wont work without this
+    runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.6")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
 }
 
 tasks.jar {
@@ -196,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Folia",
+            "Implementation-Title" to "Foldenor",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Folia",
+            "Specification-Title" to "Foldenor",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Paper Team",
-            "Brand-Id" to "papermc:folia",
-            "Brand-Name" to "Folia",
+            "Specification-Vendor" to "Edenor",
+            "Brand-Id" to "edenor:foldenor",
+            "Brand-Name" to "Foldenor",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
@@ -292,6 +_,7 @@
         // vendor.set(JvmVendorSpec.JETBRAINS)
     })
     //jvmArgs("-XX:+AllowEnhancedClassRedefinition")
+    jvmArgs("--add-modules=jdk.incubator.vector")
 
     if (rootProject.childProjects["test-plugin"] != null) {
         val testPluginJar = rootProject.project(":test-plugin").tasks.jar.flatMap { it.archiveFile }
