--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -18,7 +_,7 @@
     static {
         String bukkitVersion = "Unknown-Version";
         String apiVersion = null;
-        try (final InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("apiVersioning.json")) {
+        try (final InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/dev.edenor.foldenor/foldenor-api/pom.properties")) { // Foldenor
             if (stream == null) {
                 throw new IOException("apiVersioning.json not found in classpath");
             }
