--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -56,6 +_,18 @@
     }
     // CraftBukkit end
 
+    // Pufferfish start - Improve container checking with a bitset
+    private gg.airplane.structs.ItemListWithBitset optimizedItems;
+    @Override
+    public Container.OptionalState hasEmptySlot() {
+        return this.optimizedItems.isSlotFilled() ? Container.OptionalState.TRUE : Container.OptionalState.FALSE;
+    }
+
+    @Override
+    public Container.OptionalState isCompletelyEmpty() {
+        return this.optimizedItems.isCompletelyEmpty() ? Container.OptionalState.TRUE : Container.OptionalState.FALSE;
+    }
+    // Pufferfish end - Improve container checking with a bitset
     private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
     public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
         @Override
@@ -87,6 +_,10 @@
 
     public BarrelBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.BARREL, pos, blockState);
+        // Pufferfish start - Improve container checking with a bitset
+        this.optimizedItems = new gg.airplane.structs.ItemListWithBitset(this.getContainerSize());
+        this.items = this.optimizedItems.nonNullList;
+        // Pufferfish end - Improve container checking with a bitset
     }
 
     @Override
@@ -100,7 +_,10 @@
     @Override
     protected void loadAdditional(ValueInput input) {
         super.loadAdditional(input);
-        this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
+        // Pufferfish start - Improve container checking with a bitset
+        this.optimizedItems = new gg.airplane.structs.ItemListWithBitset(this.getContainerSize());
+        this.items = this.optimizedItems.nonNullList;
+        // Pufferfish end - Improve container checking with a bitset
         if (!this.tryLoadLootTable(input)) {
             ContainerHelper.loadAllItems(input, this.items);
         }
@@ -118,7 +_,10 @@
 
     @Override
     protected void setItems(NonNullList<ItemStack> items) {
-        this.items = items;
+        // Pufferfish start - Improve container checking with a bitset
+        this.optimizedItems = gg.airplane.structs.ItemListWithBitset.fromList(items);
+        this.items = this.optimizedItems.nonNullList;
+        // Pufferfish end - Improve container checking with a bitset
     }
 
     @Override
