--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -132,10 +_,12 @@
         return (Brain<Hoglin>) super.getBrain();
     }
 
+    private int behaviorTick = 0; // Pufferfish
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("hoglinBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Pufferfish
         this.getBrain().tick(level, this);
         profiler.pop();
         HoglinAi.updateActivity(this);
