--- a/net/minecraft/world/entity/ConversionType.java
+++ b/net/minecraft/world/entity/ConversionType.java
@@ -37,7 +_,7 @@
             }
 
             if (params.keepEquipment()) {
-                for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+                for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
                     ItemStack itemStack = from.getItemBySlot(slot);
                     if (!itemStack.isEmpty()) {
                         to.setItemSlot(slot, itemStack.copyAndClear());
