--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -28,6 +_,8 @@
 import java.util.function.BiConsumer;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -256,7 +_,7 @@
     public double xo;
     public double yo;
     public double zo;
-    private Vec3 position;
+    public Vec3 position; // Gale - JettPack - optimize sun burn tick - private -> public
     private BlockPos blockPosition;
     private ChunkPos chunkPosition;
     private Vec3 deltaMovement = Vec3.ZERO;
@@ -367,12 +_,16 @@
     public boolean fixedPose = false; // Paper - Expand Pose API
     private final int despawnTime; // Paper - entity despawn time limit
     public int totalEntityAge; // Paper - age-like counter for all entities
+    private int lastTickTicks; // Leaf - Rewrite entity despawn time
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
     // Paper start - EAR 2
     public final boolean defaultActivationState;
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public boolean activatedPriorityReset = false; // Pufferfish - DAB
+    public int activatedPriority = FoldenorConfig.maximumActivationPrio; // Pufferfish - DAB (golf score)
+    private boolean boundingBoxChanged = false; // Gale - VMP - skip entity move if movement is zero
 
     public void inactiveTick() {
     }
@@ -392,6 +_,7 @@
     private int sectionY = Integer.MIN_VALUE;
     private int sectionZ = Integer.MIN_VALUE;
     private boolean updatingSectionStatus;
+    private final boolean shouldSkipBaseDespawnCheck = this instanceof net.minecraft.world.entity.projectile.ThrowableProjectile; // Leaf - Rewrite entity despawn time
 
     @Override
     public final boolean moonrise$isHardColliding() {
@@ -580,6 +_,7 @@
         this.setPos(0.0, 0.0, 0.0);
         this.eyeHeight = this.dimensions.eyeHeight();
         this.despawnTime = level == null || type == EntityType.PLAYER ? -1 : level.paperConfig().entities.spawning.despawnTime.getOrDefault(type, io.papermc.paper.configuration.type.number.IntOr.Disabled.DISABLED).or(-1); // Paper - entity despawn time limit
+        this.updateLastTick(); // Leaf - Rewrite entity despawn time
     }
 
     public boolean isColliding(BlockPos pos, BlockState state) {
@@ -848,15 +_,57 @@
     }
 
     public void tick() {
+        // Leaf start - Rewrite entity despawn time
+        if (!this.shouldSkipBaseDespawnCheck && this.detectDespawnTime()) {
+            return;
+        }
+        /*
         // Paper start - entity despawn time limit
         if (this.despawnTime >= 0 && this.totalEntityAge >= this.despawnTime) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
             return;
         }
         // Paper end - entity despawn time limit
+        */
+        // Leaf end - Rewrite entity despawn time
         this.baseTick();
     }
 
+    // Leaf start - Rewrite entity despawn time
+    protected final boolean detectDespawnTime() {
+        if (this.despawnTime >= 0) {
+            int missedTicks = this.calculateMissedTicks();
+            if (missedTicks > 1) {
+                if ((this.totalEntityAge += missedTicks) >= this.despawnTime) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                    return true;
+                }
+            } else {
+                if (this.totalEntityAge >= this.despawnTime) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private int calculateMissedTicks() {
+        var level = this.level();
+        if (level != null) {
+            return Math.toIntExact(level.getRedstoneGameTime() - this.lastTickTicks);
+        }
+        return 0;
+    }
+
+    public void updateLastTick() {
+        var level = this.level();
+        if (level != null) {
+            this.lastTickTicks = Math.toIntExact(level.getRedstoneGameTime());
+        }
+    }
+    // Leaf end - Rewrite entity despawn time
+
     // CraftBukkit start
     public void postTick() {
         // No clean way to break out of ticking once the entity has been copied to a new world, so instead we move the portalling later in the tick cycle
@@ -916,7 +_,11 @@
 
         this.checkBelowWorld();
         if (!this.level().isClientSide()) {
-            this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0 && !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE));
+            } else {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            }
         }
 
         this.firstTick = false;
@@ -1117,28 +_,15 @@
         return this.onGround;
     }
 
-    // Paper start - detailed watchdog information
-    public final Object posLock = new Object(); // Paper - log detailed entity tick information
-
-    @Nullable
-    private Vec3 moveVector;
-    private double moveStartX;
-    private double moveStartY;
-    private double moveStartZ;
-    // Paper end - detailed watchdog information
-
     public void move(MoverType type, Vec3 movement) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.boundingBoxChanged && movement.equals(Vec3.ZERO)) {
+            return;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         final Vec3 originalMovement = movement; // Paper - Expose pre-collision velocity
-        // Paper start - detailed watchdog information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
-        synchronized (this.posLock) {
-            this.moveStartX = this.getX();
-            this.moveStartY = this.getY();
-            this.moveStartZ = this.getZ();
-            this.moveVector = movement;
-        }
         try {
-        // Paper end - detailed watchdog information
         if (this.noPhysics) {
             this.setPos(this.getX() + movement.x, this.getY() + movement.y, this.getZ() + movement.z);
             this.horizontalCollision = false;
@@ -1175,7 +_,7 @@
             }
             // Paper end
 
-            movement = this.maybeBackOffFromEdge(movement, type);
+            if (!FoldenorConfig.optimizePlayerMovementProcessing) movement = this.maybeBackOffFromEdge(movement, type); // Leaf - Optimise player movement checks
             Vec3 vec3 = this.collide(movement);
             double d = vec3.lengthSqr();
             if (d > 1.0E-7 || movement.lengthSqr() - d < 1.0E-7) {
@@ -1268,9 +_,6 @@
         }
         // Paper start - detailed watchdog information
         } finally {
-            synchronized (this.posLock) { // Paper
-                this.moveVector = null;
-            } // Paper
         }
         // Paper end - detailed watchdog information
     }
@@ -2195,10 +_,18 @@
 
     @Deprecated
     public float getLightLevelDependentMagicValue() {
-        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ())
-            ? this.level().getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ()))
-            : 0.0F;
-    }
+        return this.getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())); // Gale - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
+    }
+
+    // Gale start - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
+    /**
+       * @deprecated
+       */
+    @Deprecated
+    public float getLightLevelDependentMagicValue(BlockPos pos) {
+        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ()) ? this.level.getLightLevelDependentMagicValue(pos) : 0.0F;
+    }
+    // Gale end - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
 
     public void absSnapTo(double x, double y, double z, float yRot, float xRot) {
         this.absSnapTo(x, y, z);
@@ -2310,8 +_,7 @@
     }
 
     public void push(Entity entity) {
-        if (!this.isPassengerOfSameVehicle(entity)) {
-            if (!entity.noPhysics && !this.noPhysics) {
+        if (!entity.noPhysics && !this.noPhysics && !this.isPassengerOfSameVehicle(entity)) { // Patina - (Akarin) Swaps the predicate order of collision
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper - Collision option for requiring a player participant
                 double d = entity.getX() - this.getX();
                 double d1 = entity.getZ() - this.getZ();
@@ -2337,7 +_,6 @@
                         entity.push(d, 0.0, d1);
                     }
                 }
-            }
         }
     }
 
@@ -2597,6 +_,12 @@
             if (this.maxAirTicks != this.getDefaultMaxAirSupply()) {
                 output.putInt("Bukkit.MaxAirSupply", this.getMaxAirSupply());
             }
+            // Leaf start - Rewrite entity despawn time
+            int missedTicks = this.calculateMissedTicks();
+            if (missedTicks > 1) {
+                this.totalEntityAge += missedTicks;
+            }
+            // Leaf end - Rewrite entity despawn time
             output.putInt("Spigot.ticksLived", this.totalEntityAge); // Paper
             // CraftBukkit end
             output.storeNullable("CustomName", ComponentSerialization.CODEC, this.getCustomName());
@@ -3173,7 +_,7 @@
     }
 
     private boolean attemptToShearEquipment(Player player, InteractionHand hand, ItemStack stack, Mob mob) {
-        for (EquipmentSlot equipmentSlot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot equipmentSlot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             ItemStack itemBySlot = mob.getItemBySlot(equipmentSlot);
             Equippable equippable = itemBySlot.get(DataComponents.EQUIPPABLE);
             if (equippable != null
@@ -4896,7 +_,7 @@
         return this;
     }
 
-    private @Nullable Entity teleportCrossDimension(ServerLevel oldLevel, ServerLevel newLevel, TeleportTransition teleportTransition) {
+    protected @Nullable Entity teleportCrossDimension(ServerLevel oldLevel, ServerLevel newLevel, TeleportTransition teleportTransition) {
         List<Entity> passengers = this.getPassengers();
         List<Entity> list = new ArrayList<>(passengers.size());
         this.ejectPassengers();
@@ -4923,13 +_,13 @@
             entityx.restoreFrom(this);
             this.removeAfterChangingDimensions();
             entityx.teleportSetPosition(PositionMoveRotation.of(this), PositionMoveRotation.of(teleportTransition), teleportTransition.relatives());
-            if (this.inWorld) newLevel.addDuringTeleport(entityx); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
+            //if (this.inWorld) newLevel.addDuringTeleport(entityx); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
 
             for (Entity entity2 : list) {
                 entity2.startRiding(entityx, true, false);
             }
 
-            newLevel.resetEmptyTime();
+            //newLevel.resetEmptyTime();
             teleportTransition.postTeleportTransition().onTransition(entityx);
             this.teleportSpectators(teleportTransition, oldLevel);
             profilerFiller.pop();
@@ -5274,6 +_,11 @@
     }
 
     public final void setBoundingBox(AABB bb) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.bb.equals(bb)) {
+            this.boundingBoxChanged = true;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
                 minY = bb.minY,
@@ -5763,9 +_,7 @@
 
     public void setDeltaMovement(Vec3 deltaMovement) {
         if (deltaMovement.isFinite()) {
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.deltaMovement = deltaMovement;
-            } // Paper - detailed watchdog information
         }
     }
 
@@ -5867,9 +_,7 @@
         // Paper end - Block invalid positions and bounding box
         boolean posChanged = this.position.x != x || this.position.y != y || this.position.z != z;
         if (posChanged) { // Folia - region threading
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.position = new Vec3(x, y, z);
-            } // Paper - detailed watchdog information
             int floor = Mth.floor(x);
             int floor1 = Mth.floor(y);
             int floor2 = Mth.floor(z);
