--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -256,7 +_,7 @@
     public double xo;
     public double yo;
     public double zo;
-    private Vec3 position;
+    public Vec3 position; // Gale - JettPack - optimize sun burn tick - private -> public
     private BlockPos blockPosition;
     private ChunkPos chunkPosition;
     private Vec3 deltaMovement = Vec3.ZERO;
@@ -373,6 +_,9 @@
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public boolean activatedPriorityReset = false; // Pufferfish - DAB
+    public int activatedPriority = FoldenorConfig.maximumActivationPrio; // Pufferfish - DAB (golf score)
+    private boolean boundingBoxChanged = false; // Gale - VMP - skip entity move if movement is zero
 
     public void inactiveTick() {
     }
@@ -916,7 +_,11 @@
 
         this.checkBelowWorld();
         if (!this.level().isClientSide()) {
-            this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0 && !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE));
+            } else {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            }
         }
 
         this.firstTick = false;
@@ -1128,6 +_,11 @@
     // Paper end - detailed watchdog information
 
     public void move(MoverType type, Vec3 movement) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.boundingBoxChanged && movement.equals(Vec3.ZERO)) {
+            return;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         final Vec3 originalMovement = movement; // Paper - Expose pre-collision velocity
         // Paper start - detailed watchdog information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
@@ -1175,7 +_,7 @@
             }
             // Paper end
 
-            movement = this.maybeBackOffFromEdge(movement, type);
+            if (!FoldenorConfig.optimizePlayerMovementProcessing) movement = this.maybeBackOffFromEdge(movement, type); // Leaf - Optimise player movement checks
             Vec3 vec3 = this.collide(movement);
             double d = vec3.lengthSqr();
             if (d > 1.0E-7 || movement.lengthSqr() - d < 1.0E-7) {
@@ -2195,10 +_,18 @@
 
     @Deprecated
     public float getLightLevelDependentMagicValue() {
-        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ())
-            ? this.level().getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ()))
-            : 0.0F;
-    }
+        return this.getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())); // Gale - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
+    }
+
+    // Gale start - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
+    /**
+       * @deprecated
+       */
+    @Deprecated
+    public float getLightLevelDependentMagicValue(BlockPos pos) {
+        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ()) ? this.level.getLightLevelDependentMagicValue(pos) : 0.0F;
+    }
+    // Gale end - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
 
     public void absSnapTo(double x, double y, double z, float yRot, float xRot) {
         this.absSnapTo(x, y, z);
@@ -2310,8 +_,7 @@
     }
 
     public void push(Entity entity) {
-        if (!this.isPassengerOfSameVehicle(entity)) {
-            if (!entity.noPhysics && !this.noPhysics) {
+        if (!entity.noPhysics && !this.noPhysics && !this.isPassengerOfSameVehicle(entity)) { // Patina - (Akarin) Swaps the predicate order of collision
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper - Collision option for requiring a player participant
                 double d = entity.getX() - this.getX();
                 double d1 = entity.getZ() - this.getZ();
@@ -2337,7 +_,6 @@
                         entity.push(d, 0.0, d1);
                     }
                 }
-            }
         }
     }
 
@@ -3171,7 +_,7 @@
     }
 
     private boolean attemptToShearEquipment(Player player, InteractionHand hand, ItemStack stack, Mob mob) {
-        for (EquipmentSlot equipmentSlot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot equipmentSlot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             ItemStack itemBySlot = mob.getItemBySlot(equipmentSlot);
             Equippable equippable = itemBySlot.get(DataComponents.EQUIPPABLE);
             if (equippable != null
@@ -4894,7 +_,7 @@
         return this;
     }
 
-    private @Nullable Entity teleportCrossDimension(ServerLevel oldLevel, ServerLevel newLevel, TeleportTransition teleportTransition) {
+    protected @Nullable Entity teleportCrossDimension(ServerLevel oldLevel, ServerLevel newLevel, TeleportTransition teleportTransition) {
         List<Entity> passengers = this.getPassengers();
         List<Entity> list = new ArrayList<>(passengers.size());
         this.ejectPassengers();
@@ -4921,13 +_,13 @@
             entityx.restoreFrom(this);
             this.removeAfterChangingDimensions();
             entityx.teleportSetPosition(PositionMoveRotation.of(this), PositionMoveRotation.of(teleportTransition), teleportTransition.relatives());
-            if (this.inWorld) newLevel.addDuringTeleport(entityx); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
+            //if (this.inWorld) newLevel.addDuringTeleport(entityx); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
 
             for (Entity entity2 : list) {
                 entity2.startRiding(entityx, true, false);
             }
 
-            newLevel.resetEmptyTime();
+            //newLevel.resetEmptyTime();
             teleportTransition.postTeleportTransition().onTransition(entityx);
             this.teleportSpectators(teleportTransition, oldLevel);
             profilerFiller.pop();
@@ -5272,6 +_,11 @@
     }
 
     public final void setBoundingBox(AABB bb) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.bb.equals(bb)) {
+            this.boundingBoxChanged = true;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
                 minY = bb.minY,
