--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -94,11 +_,29 @@
             Optional<BlockPos> pos = leashData.delayedLeashInfo.right();
             if (leashUuid.isPresent()) {
                 Entity leasher = serverLevel.getEntity(leashUuid.get());
+                // Luminol start - Fix off region leashing
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(leasher)) {
+                    entity.forceDrops = true;
+                    entity.spawnAtLocation(serverLevel, Items.LEAD);
+                    entity.forceDrops = false;
+                    entity.setLeashData(null);
+                    return;
+                }
+                // Luminol end
                 if (leasher != null) {
                     setLeashedTo(entity, leasher, true);
                     return;
                 }
             } else if (pos.isPresent()) {
+                // Luminol start - Fix off region leashing
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverLevel, pos.get())) {
+                    entity.forceDrops = true;
+                    entity.spawnAtLocation(serverLevel, Items.LEAD);
+                    entity.forceDrops = false;
+                    entity.setLeashData(null);
+                    return;
+                }
+                // Luminol end
                 setLeashedTo(entity, LeashFenceKnotEntity.getOrCreateKnot(serverLevel, pos.get()), true);
                 return;
             }
