--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -11,6 +_,8 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.function.Predicate;
+import net.edenor.foldenor.config.FoldenorConfig;
+import net.edenor.foldenor.util.FoldenorUtils;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Vec3i;
@@ -141,7 +_,7 @@
     private final Sensing sensing;
     private DropChances dropChances = DropChances.DEFAULT;
     private boolean canPickUpLoot = false;
-    public boolean persistenceRequired = false;
+    public boolean persistenceRequired = false; // Let Me Despawn - make public
     private final Map<PathType, Float> pathfindingMalus = Maps.newEnumMap(PathType.class);
     public Optional<ResourceKey<LootTable>> lootTable = Optional.empty();
     public long lootTableSeed;
@@ -216,10 +_,12 @@
     public void inactiveTick() {
         super.inactiveTick();
         if (!this.aware) return; // Paper - Do not tick AI for inactive unaware mobs
-        if (this.goalSelector.inactiveTick()) {
+        // Paper start - Skip AI during inactive ticks for non-aware mobs
+        // Paper end - Skip AI during inactive ticks for non-aware mobs
+        if (this.goalSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priroity
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priroity
             this.targetSelector.tick();
         }
     }
@@ -381,7 +_,7 @@
         if (this.xpReward > 0) {
             int result = this.xpReward;
 
-            for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+            for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
                 if (slot.canIncreaseExperience()) {
                     ItemStack item = this.getItemBySlot(slot);
                     if (!item.isEmpty() && this.dropChances.byEquipment(slot) <= 1.0F) {
@@ -601,14 +_,28 @@
         }
     }
 
+    // Gale start - JettPack - optimize sun burn tick - cache eye blockpos
+    private BlockPos cached_eye_blockpos;
+    private int cached_position_hashcode;
+    // Gale end - JettPack - optimize sun burn tick - cache eye blockpos
+
     public boolean isSunBurnTick() {
         if (!this.level().isClientSide() && this.level().environmentAttributes().getValue(EnvironmentAttributes.MONSTERS_BURN, this.position())) {
-            float br = this.getLightLevelDependentMagicValue();
-            BlockPos roundedPos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
+            // Gale start - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
+            int positionHashCode = this.position.hashCode();
+            if (this.cached_position_hashcode != positionHashCode) {
+                this.cached_eye_blockpos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
+                this.cached_position_hashcode = positionHashCode;
+            }
+
+            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue(cached_eye_blockpos); // Pass BlockPos to getBrightness
+
+            // Check brightness first
+            if (lightLevelDependentMagicValue <= 0.5F) return false;
+            if (this.random.nextFloat() * 30.0F >= (lightLevelDependentMagicValue - 0.4F) * 2.0F) return false;
+            // Gale end - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
             boolean isInNonBurnableBlock = this.isInWaterOrRain() || this.isInPowderSnow || this.wasInPowderSnow;
-            if (br > 0.5F && this.random.nextFloat() * 30.0F < (br - 0.4F) * 2.0F && !isInNonBurnableBlock && this.level().canSeeSky(roundedPos)) {
-                return true;
-            }
+            return !isInNonBurnableBlock && this.level().canSeeSky(this.cached_eye_blockpos); // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
         }
 
         return false;
@@ -677,6 +_,7 @@
         this.setItemSlot(slot, itemStack);
         this.setGuaranteedDrop(slot);
         this.persistenceRequired = true;
+        FoldenorUtils.setPersistence(this, slot);
     }
 
     protected boolean canShearEquipment(final Player player) {
@@ -779,9 +_,13 @@
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) { // Paper - allow changing despawnInPeaceful
+            // Let Me Despawn start
+            if (FoldenorConfig.lmdEnabled && FoldenorUtils.pickedItems)
+                FoldenorUtils.dropEquipmentOnDiscard(this);
+            // Let Me Despawn end
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity player = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
+            Entity player = this.canvas$optimizeNearbyPlayerLookups(this.level(), this, -1.0); // Paper - Affects Spawning API // Canvas - faster player lookupsI
             if (player != null) {
                 // Paper start - Configurable despawn distances
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
@@ -793,11 +_,19 @@
                 final double distanceSquared = dxSqr + dzSqr + dySqr;
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                    // Let Me Despawn start
+                    if (FoldenorConfig.lmdEnabled && FoldenorUtils.pickedItems)
+                        FoldenorUtils.dropEquipmentOnDiscard(this);
+                    // Let Me Despawn end
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 }
 
                 if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
                     if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                        // Let Me Despawn start
+                        if (FoldenorConfig.lmdEnabled && FoldenorUtils.pickedItems)
+                            FoldenorUtils.dropEquipmentOnDiscard(this);
+                        // Let Me Despawn end
                         this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     }
                 } else {
@@ -810,6 +_,19 @@
         }
     }
 
+    // Canvas start - optimize nearby player lookups
+    private Player canvas$optimizeNearbyPlayerLookups(Level instance, Entity entity, double maxDistance) {
+        final Player closestPlayer = instance.getNearestPlayer(entity, this.getType().getCategory().getDespawnDistance());
+        if (closestPlayer != null) {
+            return closestPlayer;
+        } else {
+            final List<? extends Player> players = this.level().players();
+            if (players.isEmpty()) return null;
+            return players.get(0);
+        }
+    }
+    // Canvas end
+
     @Override
     protected final void serverAiStep() {
         this.noActionTime++;
@@ -829,16 +_,20 @@
         int idBasedTickCount = this.tickCount + this.getId();
         if (idBasedTickCount % 2 != 0 && this.tickCount > 1) {
             profiler.push("targetSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
             this.targetSelector.tickRunningGoals(false);
             profiler.pop();
             profiler.push("goalSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
             this.goalSelector.tickRunningGoals(false);
             profiler.pop();
         } else {
             profiler.push("targetSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
             this.targetSelector.tick();
             profiler.pop();
             profiler.push("goalSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
             this.goalSelector.tick();
             profiler.pop();
         }
@@ -1061,7 +_,7 @@
     protected void dropCustomDeathLoot(final ServerLevel level, final DamageSource source, final boolean killedByPlayer) {
         super.dropCustomDeathLoot(level, source, killedByPlayer);
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             if (this.shouldSkipLoot(slot)) continue; // Paper
             ItemStack itemStack = this.getItemBySlot(slot);
             float dropChance = this.dropChances.byEquipment(slot);
@@ -1105,7 +_,7 @@
     public Set<EquipmentSlot> dropPreservedEquipment(final ServerLevel level, final Predicate<ItemStack> shouldDrop) {
         Set<EquipmentSlot> slotsPreventedFromDropping = new HashSet<>();
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             ItemStack itemStack = this.getItemBySlot(slot);
             if (!itemStack.isEmpty()) {
                 if (!shouldDrop.test(itemStack)) {
@@ -1235,7 +_,7 @@
     protected void populateDefaultEquipmentEnchantments(final ServerLevelAccessor level, final RandomSource random, final DifficultyInstance localDifficulty) {
         this.enchantSpawnedWeapon(level, random, localDifficulty);
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             if (slot.getType() == EquipmentSlot.Type.HUMANOID_ARMOR) {
                 this.enchantSpawnedArmor(level, random, slot, localDifficulty);
             }
@@ -1647,7 +_,7 @@
     protected void postRemoveAfterChangingDimensions() {
         super.postRemoveAfterChangingDimensions();
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             ItemStack itemStack = this.getItemBySlot(slot);
             if (!itemStack.isEmpty()) {
                 itemStack.setCount(0);
