--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -259,7 +_,7 @@
     }
 
     private static boolean hasMapInvisibilityItemEquipped(Player player) {
-        for (EquipmentSlot equipmentSlot : EquipmentSlot.values()) {
+        for (EquipmentSlot equipmentSlot : EquipmentSlot.VALUES_ARRAY) { // Gale - JettPack - reduce array allocations
             if (equipmentSlot != EquipmentSlot.MAINHAND
                 && equipmentSlot != EquipmentSlot.OFFHAND
                 && player.getItemBySlot(equipmentSlot).is(ItemTags.MAP_INVISIBILITY_EQUIPMENT)) {
