--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -318,7 +_,9 @@
         new org.bukkit.event.vehicle.VehicleUpdateEvent(vehicle).callEvent();
 
         if (!from.equals(to)) {
+            this.blockTeleportAsync = true; // Luminol - Prevent teleprotAsync calls during move events
             new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, from, to).callEvent();
+            this.blockTeleportAsync = false; // Luminol - Prevent teleprotAsync calls during move events
         }
         // CraftBukkit end
         this.updateFluidInteraction();
@@ -403,6 +_,7 @@
 
     protected void moveAlongTrack(final ServerLevel level) {
         this.behavior.moveAlongTrack(level);
+        if (this.getVehicle() == null && this.isVehicle()) this.repositionAllPassengers(false); // Foldenor - update passengers after track movement
     }
 
     protected void comeOffTrack(final ServerLevel level) {
