--- a/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java
@@ -24,6 +_,7 @@
     private static final boolean DEFAULT_ENABLED = true;
     private boolean enabled = true;
     private boolean consumedItemThisFrame = false;
+    public int pickupImmunity = 0; // Gale - EMC - reduce hopper item checks
 
     public MinecartHopper(EntityType<? extends MinecartHopper> type, Level level) {
         super(type, level);
@@ -152,4 +_,11 @@
     }
     // Paper end
 
+    // Gale start - EMC - reduce hopper item checks
+    private long tickAttempts = 0;
+    @Override
+    public long getAndIncrementAttemptCounter() {
+        return tickAttempts++;
+    }
+    // Gale end EMC - - reduce hopper item checks
 }
