--- a/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
@@ -27,7 +_,7 @@
         } else {
             boolean foundPackedIce = false;
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) {
                 if (direction != Direction.DOWN && level.getBlockState(origin.relative(direction)).is(Blocks.PACKED_ICE)) {
                     foundPackedIce = true;
                     break;
@@ -52,7 +_,7 @@
                         );
                         BlockState placeState = level.getBlockState(placePos);
                         if (placeState.isAir() || placeState.is(Blocks.WATER) || placeState.is(Blocks.PACKED_ICE) || placeState.is(Blocks.ICE)) {
-                            for (Direction directionx : Direction.values()) {
+                            for (Direction directionx : Direction.VALUES) {
                                 BlockState relativeBlockState = level.getBlockState(placePos.relative(directionx));
                                 if (relativeBlockState.is(Blocks.BLUE_ICE)) {
                                     level.setBlock(placePos, Blocks.BLUE_ICE.defaultBlockState(), Block.UPDATE_CLIENTS);
