--- a/net/minecraft/world/level/levelgen/feature/GeodeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/GeodeFeature.java
@@ -41,7 +_,7 @@
         int maxGenOffset = config.maxGenOffset;
         List<Pair<BlockPos, Integer>> points = Lists.newLinkedList();
         int numPoints = config.distributionPoints.sample(random);
-        WorldgenRandom random1 = new WorldgenRandom(new LegacyRandomSource(level.getSeed()));
+        WorldgenRandom random1 = new net.edenor.foldenor.secureseed.WorldgenCryptoRandom(0, 0, net.edenor.foldenor.secureseed.Globals.Salt.GEODE_FEATURE, 0);
         NormalNoise noise = NormalNoise.create(random1, -4, 1.0);
         List<BlockPos> crackPoints = Lists.newLinkedList();
         double crackSizeAdjustment = (double)numPoints / config.outerWallDistance.maxInclusive();
