--- a/net/minecraft/world/level/levelgen/feature/TreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
@@ -211,7 +_,7 @@
 
                     shape.fill(posx.getX() - bounds.minX(), posx.getY() - bounds.minY(), posx.getZ() - bounds.minZ());
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) {
                         neighborPos.setWithOffset(posx, direction);
                         if (bounds.isInside(neighborPos)) {
                             int xInShape = neighborPos.getX() - bounds.minX();
