--- a/net/minecraft/world/level/levelgen/feature/TreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
@@ -205,7 +_,7 @@
 
                     discreteVoxelShape.fill(blockPos1.getX() - box.minX(), blockPos1.getY() - box.minY(), blockPos1.getZ() - box.minZ());
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) {
                         mutableBlockPos.setWithOffset(blockPos1, direction);
                         if (box.isInside(mutableBlockPos)) {
                             int i3 = mutableBlockPos.getX() - box.minX();
