--- a/net/minecraft/world/level/levelgen/feature/VinesFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
@@ -22,7 +_,7 @@
         if (!level.isEmptyBlock(origin)) {
             return false;
         } else {
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) {
                 if (direction != Direction.DOWN && VineBlock.isAcceptableNeighbour(level, origin.relative(direction), direction)) {
                     level.setBlock(origin, Blocks.VINE.defaultBlockState().setValue(VineBlock.getPropertyForFace(direction), true), Block.UPDATE_CLIENTS);
                     return true;
