--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -14,6 +_,8 @@
 import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
 import net.kyori.adventure.pointer.PointersSupplier;
 import net.kyori.adventure.util.TriState;
 import net.md_5.bungee.api.chat.BaseComponent;
@@ -285,11 +_,24 @@
 
     @Override
     public boolean teleport(Location location) {
+        if (true)
+            try {
+                this.server.getLogger().warning("Trying to teleport entity using redirection from teleport to teleportAsync. Please use teleportAsync instead of teleport!");
+                this.teleportAsync(location, TeleportCause.PLUGIN, new io.papermc.paper.entity.TeleportFlag[0]).get();
+                return true;
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
         return this.teleport(location, TeleportCause.PLUGIN);
     }
 
     @Override
     public boolean teleport(Location location, TeleportCause cause) {
+        if (true) {
+            this.server.getLogger().warning("Trying to teleport entity using redirection from teleport to teleportAsync. Please use teleportAsync instead of teleport!");
+            this.teleportAsync(location, cause);
+            return true;
+        }
         return teleport(location, cause, new TeleportFlag[0]);
     }
 
@@ -306,6 +_,16 @@
     protected boolean teleport0(Location location, TeleportCause cause, TeleportFlag... flags) {
         // Folia start - region threading
         if (true) {
+            this.server.getLogger().warning("Trying to teleport entity using redirection from teleport to teleportAsync. Please use teleportAsync instead of teleport!");
+            try {
+                this.teleportAsync(location, cause, flags).get();
+                return true;
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        // Folia start - region threading
+        if (false) {
             throw new UnsupportedOperationException("Must use teleportAsync while in region threading");
         }
         // Folia end - region threading
@@ -354,11 +_,27 @@
 
     @Override
     public boolean teleport(org.bukkit.entity.Entity destination) {
+        if (true)
+            try {
+                this.server.getLogger().warning("Trying to teleport entity using redirection from teleport to teleportAsync. Please use teleportAsync instead of teleport!");
+                this.teleportAsync(destination.getLocation(), TeleportCause.PLUGIN, new io.papermc.paper.entity.TeleportFlag[0]).get();
+                return true;
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
         return this.teleport(destination.getLocation());
     }
 
     @Override
     public boolean teleport(org.bukkit.entity.Entity destination, TeleportCause cause) {
+        if (true)
+            try {
+                this.server.getLogger().warning("Trying to teleport entity using redirection from teleport to teleportAsync. Please use teleportAsync instead of teleport!");
+                this.teleportAsync(destination.getLocation(), cause, new io.papermc.paper.entity.TeleportFlag[0]).get();
+                return true;
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
         return this.teleport(destination.getLocation(), cause);
     }
 
@@ -1132,6 +_,7 @@
 
         // Folia start - region threading
         java.util.concurrent.CompletableFuture<Boolean> ret = new java.util.concurrent.CompletableFuture<>();
+        if (location.getWorld() == null) { ret.complete(false); return ret; } // Canvas - don't crash the server if a plugin provides a null world
         java.util.function.Consumer<Entity> run = (Entity nmsEntity) -> {
             boolean success = nmsEntity.teleportAsync(
                 ((CraftWorld)locationClone.getWorld()).getHandle(),
