--- a/src/main/java/ca/spottedleaf/moonrise/common/list/EntityList.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/list/EntityList.java
@@ -18,9 +_,7 @@
         this.entityToIndex.defaultReturnValue(Integer.MIN_VALUE);
     }
 
-    private static final Entity[] EMPTY_LIST = new Entity[0];
-
-    private Entity[] entities = EMPTY_LIST;
+    private Entity[] entities = me.titaniumtown.ArrayConstants.emptyEntityArray; // Gale - JettPack - reduce array allocations
     private int count;
 
     public int size() {
