--- a/src/main/java/ca/spottedleaf/moonrise/common/list/IntList.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/list/IntList.java
@@ -10,9 +_,7 @@
         this.map.defaultReturnValue(Integer.MIN_VALUE);
     }
 
-    private static final int[] EMPTY_LIST = new int[0];
-
-    private int[] byIndex = EMPTY_LIST;
+    private int[] byIndex = me.titaniumtown.ArrayConstants.emptyIntArray; // Gale - JettPack - reduce array allocations
     private int count;
 
     public int size() {
