--- a/src/main/java/ca/spottedleaf/moonrise/common/list/ShortList.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/list/ShortList.java
@@ -10,9 +_,7 @@
         this.map.defaultReturnValue(Short.MIN_VALUE);
     }
 
-    private static final short[] EMPTY_LIST = new short[0];
-
-    private short[] byIndex = EMPTY_LIST;
+    private short[] byIndex = me.titaniumtown.ArrayConstants.emptyShortArray; // Gale - JettPack - reduce array allocations
     private short count;
 
     public int size() {
