--- a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
@@ -393,11 +_,13 @@
         return (Brain<HappyGhast>) super.getBrain();
     }
 
+    private int behaviorTick = 0; // Pufferfish
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         if (this.isBaby()) {
             ProfilerFiller profiler = Profiler.get();
             profiler.push("happyGhastBrain");
+            if (this.behaviorTick++ % this.activatedPriority == 0) // Pufferfish
             this.getBrain().tick(level, this);
             profiler.pop();
             profiler.push("happyGhastActivityUpdate");
