--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -633,6 +_,13 @@
                 this.emptyConstantCollisionShape = this.constantCollisionShape != null && this.constantCollisionShape.isEmpty();
                 // init caches
                 initCaches(collisionShape, true);
+                // Leaf start - optimize collision shape
+                switch (getBlock()) {
+                    case net.minecraft.world.level.block.FireBlock fireBlock -> {}
+                    case net.minecraft.world.level.block.LiquidBlock liquidBlock -> {}
+                    default -> this.constantCollisionShape = collisionShape;
+                }
+                // Leaf end - optimize collision shape
                 if (this.constantCollisionShape != null) {
                     initCaches(this.constantCollisionShape, true);
                 }
@@ -1051,7 +_,7 @@
             private static final Direction[] DIRECTIONS = Direction.values();
             private static final int SUPPORT_TYPE_COUNT = SupportType.values().length;
             protected final VoxelShape collisionShape;
-            protected final boolean largeCollisionShape;
+            protected boolean largeCollisionShape;
             private final boolean[] faceSturdy;
             protected final boolean isCollisionShapeFullBlock;
 
@@ -1067,8 +_,14 @@
                         )
                     );
                 } else {
-                    this.largeCollisionShape = Arrays.stream(Direction.Axis.values())
-                        .anyMatch(axis -> this.collisionShape.min(axis) < 0.0 || this.collisionShape.max(axis) > 1.0);
+                    // Leaf start - Remove stream in BlockBehaviour cache blockstate
+                    for (Direction.Axis axis : Direction.Axis.values()) {
+                        if (this.collisionShape.min(axis) < 0.0 || this.collisionShape.max(axis) > 1.0) {
+                            this.largeCollisionShape = true;
+                            break;
+                        }
+                    }
+                    // Leaf end - Remove stream in BlockBehaviour cache blockstate
                     this.faceSturdy = new boolean[DIRECTIONS.length * SUPPORT_TYPE_COUNT];
 
                     for (Direction direction : DIRECTIONS) {
