From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Thu, 22 Jun 2023 13:38:56 +0400
Subject: [PATCH] Add-deprecated-tag-to-getScheduler-function


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index db1944bc41ab0bae9b825c7839e8d1ee86eddff0..f1e91583a8503cd16961bd640f75f4436b17f5fd 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -748,8 +748,10 @@ public final class Bukkit {
      * Gets the scheduler for managing scheduled events.
      *
      * @return a scheduling service for this server
+     * @deprecated use getRegionScheduler() or getGlobalRegionScheduler() instead
      */
     @NotNull
+    @Deprecated
     public static BukkitScheduler getScheduler() {
         return server.getScheduler();
     }
