From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Thu, 22 Jun 2023 13:38:56 +0400
Subject: [PATCH] Add-deprecated-tag-to-getScheduler-function


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index fd5ed5d6725537c26e6139dd6242c4154b306a97..dd96e5bf98eb31e794d0719efc7cc3540218a934 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -749,8 +749,10 @@ public final class Bukkit {
      * Gets the scheduler for managing scheduled events.
      *
      * @return a scheduling service for this server
+     * @deprecated use getRegionScheduler() or getGlobalRegionScheduler() instead
      */
     @NotNull
+    @Deprecated
     public static BukkitScheduler getScheduler() {
         return server.getScheduler();
     }
