From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Thu, 22 Jun 2023 13:38:56 +0400
Subject: [PATCH] Add-deprecated-tag-to-getScheduler-function


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 92678097e5255a8c7ff7b6b302ccb99f13c6f2e6..95cc1c2a0fceb5c4ad677d962aaf6178b9c40bdd 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -727,8 +727,10 @@ public final class Bukkit {
      * Gets the scheduler for managing scheduled events.
      *
      * @return a scheduling service for this server
+     * @deprecated use getRegionScheduler() or getGlobalRegionScheduler() instead
      */
     @NotNull
+    @Deprecated
     public static BukkitScheduler getScheduler() {
         return server.getScheduler();
     }
