From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sat, 16 Sep 2023 21:05:49 +0400
Subject: [PATCH] Mark-all-plugins-as-folia-supported


diff --git a/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java b/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java
index b930bd722caf6ecc6e2d584535cdd7a8924d9e08..54c38af816e9f69ab8954b566c93bf9978317cbf 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java
@@ -232,7 +232,8 @@ public class PaperPluginMeta implements PluginMeta {
     // Folia start
     @Override
     public boolean isFoliaSupported() {
-        return this.foliaSupported;
+        return true;
+        //return this.foliaSupported;
     }
     // Folia end
 
