From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 12 Jun 2023 10:59:41 +0400
Subject: [PATCH] Enable-non-Folia-Plugins-Property


diff --git a/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java b/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java
index b930bd722caf6ecc6e2d584535cdd7a8924d9e08..e300a8c0eac6468214d31725674667bcb153ee0f 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperPluginMeta.java
@@ -11,6 +11,7 @@ import io.papermc.paper.plugin.provider.configuration.serializer.constraints.Plu
 import io.papermc.paper.plugin.provider.configuration.type.DependencyConfiguration;
 import io.papermc.paper.plugin.provider.configuration.type.PermissionConfiguration;
 import io.papermc.paper.plugin.provider.configuration.type.PluginDependencyLifeCycle;
+import org.bukkit.Bukkit;
 import org.bukkit.permissions.Permission;
 import org.bukkit.permissions.PermissionDefault;
 import org.bukkit.plugin.PluginLoadOrder;
@@ -232,6 +233,7 @@ public class PaperPluginMeta implements PluginMeta {
     // Folia start
     @Override
     public boolean isFoliaSupported() {
+        if (Bukkit.getServer().getEnableNonFoliaPlugins()) return true;
         return this.foliaSupported;
     }
     // Folia end
diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
index dc81e1a9faa54ba677bede6d7ca2414b1d7b6003..33fafdf3a27b9856ea27a673848a4d2e76153fbd 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
@@ -19,13 +19,14 @@ import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import java.util.logging.Logger;
 import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
+import org.bukkit.Bukkit;
 
 class PaperPluginProviderFactory implements PluginTypeFactory<PaperPluginParent, PaperPluginMeta> {
 
     @Override
     public PaperPluginParent build(JarFile file, PaperPluginMeta configuration, Path source) throws Exception {
         // Folia start - block plugins not marked as supported
-        if (!configuration.isFoliaSupported()) {
+        if (!configuration.isFoliaSupported() || !Bukkit.getServer().getEnableNonFoliaPlugins()) {
             throw new RuntimeException("Could not load plugin '" + configuration.getDisplayName() + "' as it is not marked as supporting Folia!");
         }
         // Folia end - block plugins not marked as supported
diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java b/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
index ad13d60eeb94a75c97b3d0696c39d834d6082bfa..ac3aa8da9ac51f88451b3f2b9b9b0c81aa5734c9 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
@@ -2,6 +2,7 @@ package io.papermc.paper.plugin.provider.type.spigot;
 
 import io.papermc.paper.plugin.provider.configuration.serializer.constraints.PluginConfigConstraints;
 import io.papermc.paper.plugin.provider.type.PluginTypeFactory;
+import org.bukkit.Bukkit;
 import org.bukkit.plugin.InvalidDescriptionException;
 import org.bukkit.plugin.PluginDescriptionFile;
 import org.yaml.snakeyaml.error.YAMLException;
@@ -18,7 +19,7 @@ class SpigotPluginProviderFactory implements PluginTypeFactory<SpigotPluginProvi
     @Override
     public SpigotPluginProvider build(JarFile file, PluginDescriptionFile configuration, Path source) throws Exception {
         // Folia start - block plugins not marked as supported
-        if (!configuration.isFoliaSupported()) {
+        if (!configuration.isFoliaSupported() || !Bukkit.getServer().getEnableNonFoliaPlugins()) {
             throw new RuntimeException("Could not load plugin '" + configuration.getDisplayName() + "' as it is not marked as supporting Folia!");
         }
         // Folia end - block plugins not marked as supported
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index d307826380c4bd47ee4602813737a11339cb8357..aa1824d0adb2504cd32ed0cdc33d0cb4deb0a2a9 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1745,6 +1745,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         return true;
     }
 
+    public boolean getEnableNonFoliaPlugins(){return true;}
+
     public void addTickable(Runnable tickable) {
         this.tickables.add(tickable);
     }
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 78ada86028bac06b62a7d70776a29705a3b11a06..49cc391aa6e94e9ac7f9138bb0a8851c5d14f068 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -452,6 +452,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
     public boolean isNetherEnabled() {
         return this.getProperties().allowNether;
     }
+    @Override
+    public boolean getEnableNonFoliaPlugins(){return this.getProperties().enableNonFoliaPlugins;}
 
     static final java.util.concurrent.atomic.AtomicInteger ASYNC_DEBUG_CHUNKS_COUNT = new java.util.concurrent.atomic.AtomicInteger(); // Paper - rewrite chunk system
 
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index 818289e831e3dad29345c43265e2efd7689bc500..f3a6ed604960438aa81fb959956a8026e93169bb 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -108,6 +108,8 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     private final DedicatedServerProperties.WorldDimensionData worldDimensionData;
     public final WorldOptions worldOptions;
 
+    public final boolean enableNonFoliaPlugins = this.get("enableNonFoliaPlugins", true);
+
     public final String rconIp; // Paper - Add rcon ip
 
     // CraftBukkit start
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 20af7484bf570d28cd9723fdbb1ec07657117367..94210bc8e7382dea11187cc1ecc293aa1da589ed 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1759,6 +1759,8 @@ public final class CraftServer implements Server {
         return this.console.usesAuthentication();
     }
 
+    public boolean getEnableNonFoliaPlugins() {return this.console.getEnableNonFoliaPlugins();}
+
     @Override
     public boolean getAllowFlight() {
         return this.console.isFlightAllowed();
