From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 10 Jul 2023 13:30:00 +0400
Subject: [PATCH] Return-rcon


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 34820b8a05cf1265f0eeb6e60803e32fbfe9e977..54d7566b8d8ed2b6cb5561a0a0e6d3b0b0054f7b 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -764,7 +764,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         // Folia end - region threading
         Waitable[] waitableArray = new Waitable[1];
         this.rconConsoleSource.prepareForCommand();
-        this.executeBlocking(() -> {
+        //this.executeBlocking(() -> {
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTaskWithoutNotify(() ->{
             // CraftBukkit start - fire RemoteServerCommandEvent
             RemoteServerCommandEvent event = new RemoteServerCommandEvent(remoteConsole, command);
             server.getPluginManager().callEvent(event);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c6b9714c801647690b1cb1677b6a296b2a4a43eb..6da85e5743670f67ec9cd9a0ec178ce22499055b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -977,7 +977,7 @@ public final class CraftServer implements Server {
             });
         } else {
             // huh?
-            throw new UnsupportedOperationException("Dispatching command for " + sender);
+            //throw new UnsupportedOperationException("Dispatching command for " + sender);
         }
     }
     // Folia end - region threading
@@ -995,7 +995,7 @@ public final class CraftServer implements Server {
             io.papermc.paper.threadedregions.RegionizedServer.ensureGlobalTickThread("Dispatching command async");
         } else {
             // huh?
-            throw new UnsupportedOperationException("Dispatching command for " + sender);
+            //throw new UnsupportedOperationException("Dispatching command for " + sender);
         }
         // Folia end - region threading
 
