From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Wed, 31 Jul 2024 14:59:46 +0400
Subject: [PATCH] Redirect-CraftScheduler-to-GlobalRegionScheduler


diff --git a/build.gradle.kts b/build.gradle.kts
index 76949e896e5a6804ba98753a1ebfba9b0697d47f..ecab96e631e7d6a0d3707ea4c4429ea79ae62740 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,3 +1,4 @@
+
 import io.papermc.paperweight.util.*
 import java.time.Instant
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
index fad85bea8643a3a88ec5c4194de7a5060e81c136..8ab51adc6eb4e98e1586786317e8ca31838fd0dc 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -2,21 +2,17 @@ package org.bukkit.craftbukkit.scheduler;
 
 import com.google.common.base.Preconditions;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
-import java.util.ArrayList;
-import java.util.Comparator;
-import java.util.Iterator;
-import java.util.List;
-import java.util.PriorityQueue;
-import java.util.concurrent.Callable;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.Executor;
-import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
+
+import java.util.*;
+import java.util.concurrent.*;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.IntUnaryOperator;
 import java.util.logging.Level;
+
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
+import org.bukkit.Bukkit;
 import org.bukkit.plugin.IllegalPluginAccessException;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.scheduler.BukkitRunnable;
@@ -111,6 +107,8 @@ public class CraftScheduler implements BukkitScheduler {
     */ // Paper end
     private static final int RECENT_TICKS;
 
+    private static final Map<Integer, ScheduledTask> taskMap = new HashMap<>();
+
     static {
         RECENT_TICKS = 30;
     }
@@ -266,7 +264,10 @@ public class CraftScheduler implements BukkitScheduler {
         }
         // Paper start
         if (!this.isAsyncScheduler) {
-            this.asyncScheduler.cancelTask(taskId);
+            //this.asyncScheduler.cancelTask(taskId);
+            if (taskMap.containsKey(taskId)) {
+                taskMap.get(taskId).cancel();
+            }
         }
         // Paper end
         CraftTask task = this.runners.get(taskId);
@@ -304,6 +305,7 @@ public class CraftScheduler implements BukkitScheduler {
             }
             if (taskPending.getTaskId() == taskId) {
                 taskPending.cancel0();
+                taskMap.remove(taskPending.getTaskId());
             }
         }
     }
@@ -313,7 +315,8 @@ public class CraftScheduler implements BukkitScheduler {
         Preconditions.checkArgument(plugin != null, "Cannot cancel tasks of null plugin");
         // Paper start
         if (!this.isAsyncScheduler) {
-            this.asyncScheduler.cancelTasks(plugin);
+            //this.asyncScheduler.cancelTasks(plugin);
+            Bukkit.getAsyncScheduler().cancelTasks(plugin);
         }
         // Paper end
         final CraftTask task = new CraftTask(
@@ -349,6 +352,7 @@ public class CraftScheduler implements BukkitScheduler {
         for (CraftTask runner : this.runners.values()) {
             if (runner.getOwner().equals(plugin)) {
                 runner.cancel0();
+                taskMap.remove(runner.getTaskId());
             }
         }
     }
@@ -514,15 +518,69 @@ public class CraftScheduler implements BukkitScheduler {
     }
 
     protected CraftTask handle(final CraftTask task, final long delay) { // Paper
-        if (true) throw new UnsupportedOperationException(); // Folia - region threading
+        //if (true) throw new UnsupportedOperationException(); // Folia - region threading
+        Plugin owner = task.getOwner();
+        /*if (owner == null) {
+            throw new IllegalCallerException("Attempt to call task without the owning plugin");
+        }*/
+        boolean repeat = task.getPeriod() != CraftTask.NO_REPEATING;
         // Paper start
         if (!this.isAsyncScheduler && !task.isSync()) {
-            this.asyncScheduler.handle(task, delay);
+            //this.asyncScheduler.handle(task, delay);
+            long millis_delay = (long) (delay / 20.0 * 1000.0);
+            ScheduledTask scheduledTask;
+            if (millis_delay <= 0) {
+                if (repeat) {
+                    scheduledTask = Bukkit.getAsyncScheduler().runAtFixedRate(owner, stask -> {
+                        task.run();
+                    }, 1, (long) (task.getPeriod() / 20.0 * 1000.0), TimeUnit.MILLISECONDS);
+                } else {
+                    scheduledTask = Bukkit.getAsyncScheduler().runNow(owner, stask -> {
+                        task.run();
+                    });
+                }
+            } else {
+                if (repeat) {
+                    scheduledTask = Bukkit.getAsyncScheduler().runAtFixedRate(owner, stask -> {
+                        task.run();
+                    }, millis_delay, (long) (task.getPeriod() / 20.0 * 1000.0), TimeUnit.MILLISECONDS);
+                } else {
+                    scheduledTask = Bukkit.getAsyncScheduler().runDelayed(owner, stask -> {
+                        task.run();
+                    }, millis_delay, TimeUnit.MILLISECONDS);
+                }
+            }
+            task.setScheduledTask(scheduledTask);
+            taskMap.put(task.getTaskId(), scheduledTask);
             return task;
         }
         // Paper end
-        task.setNextRun(this.currentTick + delay);
-        this.addTask(task);
+        //task.setNextRun(this.currentTick + delay);
+        //this.addTask(task);
+        ScheduledTask scheduledTask;
+        if (delay <= 0) {
+            if (repeat) {
+                scheduledTask = Bukkit.getGlobalRegionScheduler().runAtFixedRate(owner, stask -> {
+                    task.run();
+                }, 1L, task.getPeriod());
+            } else {
+                scheduledTask = Bukkit.getGlobalRegionScheduler().run(owner, stask -> {
+                    task.run();
+                });
+            }
+        } else {
+            if (repeat) {
+                scheduledTask = Bukkit.getGlobalRegionScheduler().runAtFixedRate(owner, stask -> {
+                    task.run();
+                }, delay, task.getPeriod());
+            } else {
+                scheduledTask = Bukkit.getGlobalRegionScheduler().runDelayed(owner, stask -> {
+                    task.run();
+                }, delay);
+            }
+        }
+        task.setScheduledTask(scheduledTask);
+        taskMap.put(task.getTaskId(), scheduledTask);
         return task;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
index 17680f112d0c7e7aee07e34477daa21ef2ddaa6f..33ba7f4f63f52a9b402bca45f28067c88408f886 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftTask.java
@@ -2,6 +2,8 @@ package org.bukkit.craftbukkit.scheduler;
 
 import java.util.function.Consumer;
 
+import co.aikar.timings.NullTimingHandler;
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
 import org.bukkit.Bukkit;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.scheduler.BukkitTask;
@@ -29,6 +31,7 @@ public class CraftTask implements BukkitTask, Runnable { // Spigot
     private final Plugin plugin;
     private final int id;
     private final long createdAt = System.nanoTime();
+    private ScheduledTask scheduledTask;
 
     CraftTask() {
         this(null, null, CraftTask.NO_REPEATING, CraftTask.NO_REPEATING);
@@ -64,7 +67,10 @@ public class CraftTask implements BukkitTask, Runnable { // Spigot
 
     @Override
     public final Plugin getOwner() {
-        return this.plugin;
+        if (this.plugin == null && this.scheduledTask != null && this.scheduledTask.getOwningPlugin() != null)
+            return this.scheduledTask.getOwningPlugin();
+        else
+            return this.plugin;
     }
 
     @Override
@@ -115,12 +121,19 @@ public class CraftTask implements BukkitTask, Runnable { // Spigot
 
     @Override
     public boolean isCancelled() {
-        return (this.period == CraftTask.CANCEL);
+        //return (this.period == CraftTask.CANCEL);
+        return scheduledTask != null ? scheduledTask.isCancelled() : (this.period == CraftTask.CANCEL);
     }
 
     @Override
     public void cancel() {
-        Bukkit.getScheduler().cancelTask(this.id);
+        //Bukkit.getScheduler().cancelTask(this.id);
+        if (this.scheduledTask != null)
+            scheduledTask.cancel();
+    }
+
+    protected void setScheduledTask(ScheduledTask scheduledTask) {
+        this.scheduledTask = scheduledTask;
     }
 
     /**
