From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sat, 16 Sep 2023 21:32:22 +0400
Subject: [PATCH] Multiply-spawn-limit-on-player-count-in-region


diff --git a/src/main/java/net/minecraft/world/level/NaturalSpawner.java b/src/main/java/net/minecraft/world/level/NaturalSpawner.java
index a60e63dce1869c41fe4ff97901b7143e54d1a090..4fc4c2a462f4654ecb7b989419563b2eff3e4336 100644
--- a/src/main/java/net/minecraft/world/level/NaturalSpawner.java
+++ b/src/main/java/net/minecraft/world/level/NaturalSpawner.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.level;
 
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.threadedregions.TickRegionScheduler;
 import it.unimi.dsi.fastutil.objects.Object2IntMap;
 import it.unimi.dsi.fastutil.objects.Object2IntMaps;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
@@ -143,7 +144,7 @@ public final class NaturalSpawner {
             SpawnCategory spawnCategory = CraftSpawnCategory.toBukkit(enumcreaturetype);
             if (CraftSpawnCategory.isValidForLimits(spawnCategory)) {
                 spawnThisTick = world.ticksPerSpawnCategory.getLong(spawnCategory) != 0 && world.getRedstoneGameTime() % world.ticksPerSpawnCategory.getLong(spawnCategory) == 0; // Folia - region threading
-                limit = world.getWorld().getSpawnLimit(spawnCategory);
+                limit = world.getWorld().getSpawnLimit(spawnCategory) * TickRegionScheduler.getCurrentRegion().getData().getRegionStats().getPlayerCount();
             }
 
             if (!spawnThisTick || limit == 0) {
