From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sat, 16 Sep 2023 21:34:57 +0400
Subject: [PATCH] Do-not-send-chunk-if-it-was-already-sent


diff --git a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
index c5df121d6194a97b20dc390698991b9c72dba538..76c6bf3e416cc2a6bc0d85716c7f3ce12aa8c96f 100644
--- a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
@@ -480,6 +480,10 @@ public class RegionizedPlayerChunkLoader {
         }
 
         private void sendChunk(final int chunkX, final int chunkZ) {
+            if (this.sentChunks.contains(CoordinateUtils.getChunkKey(chunkX, chunkZ))){
+                return;
+            }
+
             if (this.sentChunks.add(CoordinateUtils.getChunkKey(chunkX, chunkZ))) {
                 this.world.getChunkSource().chunkMap.updateChunkTracking(this.player,
                     new ChunkPos(chunkX, chunkZ), new MutableObject<>(), false, true); // unloaded, loaded
