From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Fri, 22 Dec 2023 19:52:19 +0400
Subject: [PATCH] fixup! Add-config


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 9a67b0107f4b79911b4d1e676e4d97c7ff8466a9..d9931ba2cded417ff98a35309ed5437778028be9 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -24,19 +24,24 @@ public class FoldenorConfig {
     public static int version;
     static boolean verbose;
 
+    //Network
     public static boolean sendNullEntityPackets = true;
 
-    public static boolean vanilaEndPortalTeleportation = false;
 
+    //Optimization
     public static int maxProjectileLoadsPerProjectile = 10;
 
     public static int maxProjectileLoadsPerTick = 10;
 
+    //World
     public static RegionFileFormat regionFormatName = RegionFileFormat.ANVIL;
     public static int regionFormatLinearCompressionLevel = 1;
     public static int linearFlushFrequency = 10;
     public static int linearFlushThreads = 1;
 
+    //Other
+    public static boolean vanilaEndPortalTeleportation = false;
+
     public static void init(File configFile) {
         init(configFile, true);
     }
