From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 6 Aug 2024 15:43:15 +0400
Subject: [PATCH] Purpur-Alternative-Keepalive-Handling


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index e9f5b546c8b6545c611d8e6cc3a91cb738c5cca3..373ed11dac6d42dc0c2ea4a59e74b9c503b113d1 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -37,14 +37,6 @@ public class FoldenorConfig {
 
     public static boolean useAlternateKeepAlive = false;
 
-    public static int piglinSpawnChancePersentInPortal = 100;
-
-    public static boolean useVirtualThreadForAsyncScheduler = false;
-
-    public static boolean skipMapItemUpdatesIfNoBukkitRender = true;
-
-    public static int entityActivationCheckFrequency = 20;
-
     public static void init(File configFile) {
         init(configFile, true);
     }
@@ -87,13 +79,11 @@ public class FoldenorConfig {
     static void readConfig() {
         readNetworkSettings();
 
-        readOptimizationSettings();
-
-        readMiscSettings();
-
-        projectileTimeouts();
-
-        pcaSyncProtocol();
+        try {
+            readDynamicActivationOfBrains();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
 
         try {
             config.save(CONFIG_FILE);
@@ -108,18 +98,6 @@ public class FoldenorConfig {
         useAlternateKeepAlive = getBoolean("network.use-alternate-keepalive", useAlternateKeepAlive);
     }
 
-    private static void readOptimizationSettings() {
-        piglinSpawnChancePersentInPortal = getInt("optimizations.piglin-spawn-chance-persent-in-portal", 100,
-            "Reduces piglin spawn in portal, by reducing change to spawn");
-        skipMapItemUpdatesIfNoBukkitRender = getBoolean("optimizations.skip_map_item_updates_if_no_bukkit_render", skipMapItemUpdatesIfNoBukkitRender);
-        entityActivationCheckFrequency = getInt("optimizations.entity-activation-check-frequency", 20);
-    }
-
-    private static void readMiscSettings() {
-        useVirtualThreadForAsyncScheduler = getBoolean("optimizations.use-virtual-thread-for-async-scheduler", useVirtualThreadForAsyncScheduler,
-            "Use the new Virtual Thread introduced in JDK 21 for CraftAsyncScheduler.");
-    }
-
     private static void readDynamicActivationOfBrains() throws IOException {
         dearEnabled = getBoolean("dab.enabled",  true);
         startDistance = getInt("dab.start-distance",  12,
@@ -144,37 +122,6 @@ public class FoldenorConfig {
                     }, () -> MinecraftServer.LOGGER.warn("Unknown entity \"" + name + "\"")));
     }
 
-    public static Map<String, Integer> projectileTimeouts;
-	private static void projectileTimeouts() {
-        // Set some defaults
-        getInt("entity_timeouts.SNOWBALL", -1);
-        getInt("entity_timeouts.LLAMA_SPIT", -1);
-        //"These values define a entity's maximum lifespan. If an",
-        //"entity is in this list and it has survived for longer than",
-        //"that number of ticks, then it will be removed. Setting a value to",
-        //"-1 disables this feature."
-
-        for (EntityType<?> entityType : BuiltInRegistries.ENTITY_TYPE) {
-            String type = EntityType.getKey(entityType).getPath().toUpperCase(Locale.ROOT);
-            entityType.ttl = config.getInt("entity_timeouts." + type, -1);
-        }
-    }
-
-    public static boolean pcaSyncProtocolEnabled = false;
-    public static SyncOption pcaSyncProtocolOption = SyncOption.OPS;
-    private static void pcaSyncProtocol() {
-        pcaSyncProtocolEnabled = getBoolean("pcaSyncProtocol.enabled", pcaSyncProtocolEnabled);
-        pcaSyncProtocolOption = SyncOption.valueOf(getString("pcaSyncProtocol.option", String.valueOf(pcaSyncProtocolOption)));
-    }
-
-    public enum SyncOption {
-        NOBODY,
-        BOT,
-        OPS,
-        OPS_AND_SELF,
-        EVERYONE
-    }
-
     protected static void set(String path, Object val) {
         config.addDefault(path, val);
         config.set(path, val);
diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 2bfeec07d6f419e6ecd5b2df7bc4fa5a431ccfb1..182ed96bce70e3cef9a3d2ad719eb2183b8ca8e2 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -80,6 +80,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     private long keepAliveChallenge;
     private long closedListenerTime;
     private boolean closed = false;
+    private it.unimi.dsi.fastutil.longs.LongList keepAlives = new it.unimi.dsi.fastutil.longs.LongArrayList(); // Purpur
     private int latency;
     private volatile boolean suspendFlushingOnServerThread = false;
     public final java.util.Map<java.util.UUID, net.kyori.adventure.resource.ResourcePackCallback> packCallbacks = new java.util.concurrent.ConcurrentHashMap<>(); // Paper - adventure resource pack callbacks
@@ -148,6 +149,16 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
     @Override
     public void handleKeepAlive(ServerboundKeepAlivePacket packet) {
+        // Purpur start
+        if (net.edenor.foldenor.config.FoldenorConfig.useAlternateKeepAlive) {
+            if (this.keepAlivePending && !keepAlives.isEmpty() && keepAlives.contains(packet.getId())) {
+                int ping = (int) (Util.getMillis() - packet.getId());
+                this.latency = (this.latency * 3 + ping) / 4;
+                this.keepAlivePending = false;
+                keepAlives.clear(); // we got a valid response, lets roll with it and forget the rest
+            }
+        } else
+            // Purpur end
         //PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel()); // CraftBukkit // Paper - handle ServerboundKeepAlivePacket async
         if (this.keepAlivePending && packet.getId() == this.keepAliveChallenge) {
             int i = (int) (Util.getMillis() - this.keepAliveTime);
@@ -279,6 +290,21 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         long currentTime = Util.getMillis();
         long elapsedTime = currentTime - this.keepAliveTime;
 
+        // Purpur start
+        if (net.edenor.foldenor.config.FoldenorConfig.useAlternateKeepAlive) {
+            if (elapsedTime >= 1000L) { // 1 second
+                if (this.keepAlivePending && !this.processedDisconnect && keepAlives.size() * 1000L >= KEEPALIVE_LIMIT) {
+                    this.disconnect(ServerCommonPacketListenerImpl.TIMEOUT_DISCONNECTION_MESSAGE, org.bukkit.event.player.PlayerKickEvent.Cause.TIMEOUT);
+                } else if (this.checkIfClosed(currentTime)) {
+                    this.keepAlivePending = true;
+                    this.keepAliveTime = currentTime; // hijack this field for 1 second intervals
+                    this.keepAlives.add(currentTime); // currentTime is ID
+                    this.send(new ClientboundKeepAlivePacket(currentTime));
+                }
+            }
+        } else
+            // Purpur end
+
         if (!this.isSingleplayerOwner() && elapsedTime >= 15000L) { // Paper - use vanilla's 15000L between keep alive packets
             if (this.keepAlivePending && !this.processedDisconnect && elapsedTime >= KEEPALIVE_LIMIT) { // Paper - check keepalive limit, don't fire if already disconnected
                 this.disconnect(ServerCommonPacketListenerImpl.TIMEOUT_DISCONNECTION_MESSAGE, PlayerKickEvent.Cause.TIMEOUT); // Paper - kick event cause
