From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 6 Aug 2024 16:19:01 +0400
Subject: [PATCH] Add-chance-config-for-piglin-spawn-in-nether-portal


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 373ed11dac6d42dc0c2ea4a59e74b9c503b113d1..dc3eefb624b0f8e6cb196f9aa72857ab5a369e10 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -37,6 +37,8 @@ public class FoldenorConfig {
 
     public static boolean useAlternateKeepAlive = false;
 
+    public static int piglinSpawnChancePersentInPortal = 100;
+
     public static void init(File configFile) {
         init(configFile, true);
     }
@@ -79,6 +81,8 @@ public class FoldenorConfig {
     static void readConfig() {
         readNetworkSettings();
 
+        readOptimizationSettings();
+
         try {
             readDynamicActivationOfBrains();
         } catch (IOException e) {
@@ -98,6 +102,11 @@ public class FoldenorConfig {
         useAlternateKeepAlive = getBoolean("network.use-alternate-keepalive", useAlternateKeepAlive);
     }
 
+    private static void readOptimizationSettings() {
+        piglinSpawnChancePersentInPortal = getInt("optimizations.piglin-spawn-chance-persent-in-portal", 100,
+            "Reduces piglin spawn in portal, by reducing change to spawn");
+    }
+
     private static void readDynamicActivationOfBrains() throws IOException {
         dearEnabled = getBoolean("dab.enabled",  true);
         startDistance = getInt("dab.start-distance",  12,
diff --git a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
index 88e33ae16848afe9885684537ef32e1836425fa3..f097dd5c4d2ea147e81d064a93ed084dddd0b9d6 100644
--- a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -4,6 +4,8 @@ import com.mojang.logging.LogUtils;
 import com.mojang.serialization.MapCodec;
 import java.util.Optional;
 import javax.annotation.Nullable;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.BlockUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -78,7 +80,9 @@ public class NetherPortalBlock extends Block implements Portal {
 
     @Override
     protected void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
-        if (world.spigotConfig.enableZombiePigmenPortalSpawns && world.dimensionType().natural() && world.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && random.nextInt(2000) < world.getDifficulty().getId()) { // Spigot
+        if (world.spigotConfig.enableZombiePigmenPortalSpawns && world.dimensionType().natural()
+            && world.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && random.nextInt(2000) < world.getDifficulty().getId()
+            && random.nextInt(100) <= FoldenorConfig.piglinSpawnChancePersentInPortal) { // Spigot // Foldenor
             while (world.getBlockState(pos).is((Block) this)) {
                 pos = pos.below();
             }
