From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Fri, 16 Jun 2023 12:16:53 +0400
Subject: [PATCH] Get-block-state-if-loaded


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 66ee61f27dc435e27c4792b6cf4f320b91a54c57..2df67e75f8960fe07a75e88f2f650bab18054579 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -239,7 +239,7 @@ public class CraftBlock implements Block {
 
     @Override
     public Material getType() {
-        return this.world.getBlockState(this.position).getBukkitMaterial(); // Paper - optimise getType calls
+        return this.world.getBlockStateIfLoaded(this.position).getBukkitMaterial(); // Paper - optimise getType calls
     }
 
     @Override
