From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 6 Aug 2024 19:29:22 +0400
Subject: [PATCH] SparklyPaper Skip 'MapItem#update()' if the map does not have


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 9ecfcf62f4add8968010374f46e14e15a86c9fe1..0dea335420bc0acacf9045fe0123593cbb689287 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -41,6 +41,8 @@ public class FoldenorConfig {
 
     public static boolean useVirtualThreadForAsyncScheduler = false;
 
+    public static boolean skipMapItemUpdatesIfNoBukkitRender = true;
+
     public static void init(File configFile) {
         init(configFile, true);
     }
@@ -109,6 +111,7 @@ public class FoldenorConfig {
     private static void readOptimizationSettings() {
         piglinSpawnChancePersentInPortal = getInt("optimizations.piglin-spawn-chance-persent-in-portal", 100,
             "Reduces piglin spawn in portal, by reducing change to spawn");
+        skipMapItemUpdatesIfNoBukkitRender = getBoolean("optimizations.skip_map_item_updates_if_no_bukkit_render", skipMapItemUpdatesIfNoBukkitRender);
     }
 
     private static void readMiscSettings() {
diff --git a/src/main/java/net/minecraft/world/item/MapItem.java b/src/main/java/net/minecraft/world/item/MapItem.java
index 65d1af43b8bc740a9dc77dd8d5ba99de000fe57a..9c98fa64e9875ca8e126fce55dea13b66a039b32 100644
--- a/src/main/java/net/minecraft/world/item/MapItem.java
+++ b/src/main/java/net/minecraft/world/item/MapItem.java
@@ -6,6 +6,8 @@ import com.google.common.collect.Multiset;
 import com.google.common.collect.Multisets;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.ChatFormatting;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -280,7 +282,7 @@ public class MapItem extends Item {
                     mapItemSavedData.tickCarriedBy(player, stack);
                 }
 
-                if (!mapItemSavedData.locked && (selected || entity instanceof Player && ((Player)entity).getOffhandItem() == stack)) {
+                    if (!mapItemSavedData.locked && (!FoldenorConfig.skipMapItemUpdatesIfNoBukkitRender || mapItemSavedData.mapView.getRenderers().stream().anyMatch(mapRenderer -> mapRenderer.getClass() == org.bukkit.craftbukkit.map.CraftMapRenderer.class)) && (selected || entity instanceof Player && ((Player)entity).getOffhandItem() == stack)) { // SparklyPaper - don't update maps if they don't have the CraftMapRenderer in the render list
                     this.update(world, entity, mapItemSavedData);
                 }
                 } // Folia - region threading
