From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 6 Aug 2024 19:37:22 +0400
Subject: [PATCH] Divine-lithium-ai.raid


diff --git a/src/main/java/net/minecraft/world/entity/raid/Raid.java b/src/main/java/net/minecraft/world/entity/raid/Raid.java
index 1df579db8286113f8412fe678259fb53098e57ed..bcaaf6cffad5be4d402a7c8ddade11e56a4c8b6c 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raid.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raid.java
@@ -107,6 +107,7 @@ public class Raid {
     private Raid.RaidStatus status;
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos;
+    private boolean isBarDirty; // DivineMC - lithium: ai.raid
     // Paper start
     private static final String PDC_NBT_KEY = "BukkitValues";
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry PDC_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
@@ -299,6 +300,12 @@ public class Raid {
 
     public void tick() {
         if (!this.isStopped()) {
+            // DivineMC start - lithium: ai.raid
+            if (this.isBarDirty) {
+                this.updateBossbarInternal();
+                this.isBarDirty = false;
+            }
+            // DivineMC end
             if (this.status == Raid.RaidStatus.ONGOING) {
                 boolean flag = this.active;
 
@@ -666,21 +673,24 @@ public class Raid {
 
     }
 
+    // DivineMC start - lithium: ai.raid
     public void updateBossbar() {
+        this.isBarDirty = true;
+    }
+
+    private void updateBossbarInternal() {
         this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
     }
+    // DivineMC end
 
     public float getHealthOfLivingRaiders() {
         float f = 0.0F;
-        Iterator iterator = this.groupRaiderMap.values().iterator();
-
-        while (iterator.hasNext()) {
-            Set<Raider> set = (Set) iterator.next();
+        for (Set<Raider> raiders : this.groupRaiderMap.values()) {
 
             Raider entityraider;
 
-            for (Iterator iterator1 = set.iterator(); iterator1.hasNext(); f += entityraider.getHealth()) {
-                entityraider = (Raider) iterator1.next();
+            for (Iterator<Raider> iterator1 = raiders.iterator(); iterator1.hasNext(); f += entityraider.getHealth()) {
+                entityraider = iterator1.next();;
             }
         }
 
@@ -696,7 +706,7 @@ public class Raid {
     }
 
     public void removeFromRaid(Raider entity, boolean countHealth) {
-        Set<Raider> set = (Set) this.groupRaiderMap.get(entity.getWave());
+        Set<Raider> set = this.groupRaiderMap.get(entity.getWave());
 
         if (set != null) {
             boolean flag1 = set.remove(entity);
