From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sun, 25 Jun 2023 13:50:13 +0400
Subject: [PATCH] Fix-death-broadcast-cannot-be-ended


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 511ec694f3336a5d7276d171dd2345ed2ea1acac..620c799bb9a23e97686e0102e74aa7ecc9f99040 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -633,6 +633,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         ++this.deathTime;
         if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved() && !this.broadcastedDeath) { // Folia - region threading
             this.level().broadcastEntityEvent(this, (byte) 60);
+            this.broadcastedDeath = true;
             if (!(this instanceof ServerPlayer)) this.remove(Entity.RemovalReason.KILLED); // Folia - region threading - don't remove, we want the tick scheduler to be running
             if ((this instanceof ServerPlayer)) this.unRide(); // Folia - region threading - unmount player when dead
         }
