From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 6 Aug 2024 19:38:22 +0400
Subject: [PATCH] Divine: lithium: entity.fast_elytra_check +
 entity.fast_hand_swing


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 7b745943a781906b2ef87510557aec08ee396d8e..b674ca7318095509b1268d6b2ed5f4263d8c797b 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2768,6 +2768,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected void updateSwingTime() {
+        if (!this.swinging && this.swingTime == 0) return; // DivineMC - lithium: entity.fast_hand_swing
         int i = this.getCurrentSwingDuration();
 
         if (this.swinging) {
@@ -3760,6 +3761,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected void updateFallFlying() {
+        if (!this.isFallFlying()) return; // DivineMC - lithium: entity.fast_elytra_check
         this.checkSlowFallDistance();
         if (!this.level().isClientSide) {
             if (!this.canGlide()) {
