From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Wed, 7 Aug 2024 17:07:26 +0400
Subject: [PATCH] Change '/profiler' permission


diff --git a/src/main/java/io/papermc/paper/threadedregions/commands/CommandProfiler.java b/src/main/java/io/papermc/paper/threadedregions/commands/CommandProfiler.java
index dbc6ffd8fec4570de4abdb3aa0a16d0e0cc09353..bb5bad1cd580bd7b44e2c4ed8abc840a536d72a4 100644
--- a/src/main/java/io/papermc/paper/threadedregions/commands/CommandProfiler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/commands/CommandProfiler.java
@@ -52,7 +52,7 @@ public final class CommandProfiler extends Command {
         super("profiler");
         this.setUsage("/<command> <world> <block x> <block z> <time in s> [radius, default 100 blocks]");
         this.setDescription("Reports information about server health.");
-        this.setPermission("bukkit.command.tps");
+        this.setPermission("paper.profiler");
     }
 
     @Override
