From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 12 Jun 2023 15:51:31 +0400
Subject: [PATCH] Something


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index d4a1d17d5dde68b9bc2fbae370ef2268f3b604db..aed187bfea9f7976c839bdda325bd12158fa34bf 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -240,7 +240,7 @@ public class CraftBlock implements Block {
 
     @Override
     public Material getType() {
-        return this.world.getBlockState(this.position).getBukkitMaterial(); // Paper - optimise getType calls
+        return this.world.getBlockStateIfLoaded(this.position).getBukkitMaterial(); // Paper - optimise getType calls
     }
 
     @Override
