From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Thu, 4 Jan 2024 16:39:26 +0400
Subject: [PATCH] Change-reload-config-work


diff --git a/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java b/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java
index 4188e8db6386fe1790192c9efd8c2fe256ce6418..9b49dafe2706a41200b1f2ee74f4b857346e5418 100644
--- a/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java
+++ b/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java
@@ -38,7 +38,7 @@ public class FoldenorReloadConfigCommand extends Command {
         if (!testPermission(sender)) return true;
 
         try {
-            FoldenorConfig.reload((java.io.File) DedicatedServer.getServer().options.valueOf("foldenor-settings"));
+            FoldenorConfig.readConfig();
             Command.broadcastCommandMessage(sender, text("Foldenor config reload complete.", GREEN));
             return true;
         } catch (Exception e) {
diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index c668d7e2c64379e48a12a4c93dbdf18a5ed8cbe7..085495a2d82f2a514a01602c8aece5ace57eb240 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -83,7 +83,7 @@ public class FoldenorConfig {
         Bukkit.getLogger().log(level, s);
     }
 
-    static void readConfig() {
+    public static void readConfig() {
         networkSettings();
 
         optimizationSettings();
