From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 12 Jun 2023 16:01:31 +0400
Subject: [PATCH] Check-if-block-loaded


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 8cf4428d8140fd03f0ad8fa1d148a04d8caefda8..9884ab63c042e008cbcbe77561381757b0cdab4d 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -697,7 +697,8 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
 
     @Override
     public boolean destroyBlock(BlockPos pos, boolean drop, @Nullable Entity breakingEntity, int maxUpdateDepth) {
-        BlockState iblockdata = this.getBlockState(pos);
+        BlockState iblockdata = !io.papermc.paper.util.TickThread.isTickThreadFor((ServerLevel)this, pos) ? null : this.getBlockStateIfLoaded(pos); // Folia - block updates in unloaded chunks
+        if (iblockdata == null) return false;
 
         if (iblockdata.isAir()) {
             return false;
