From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 20 Aug 2024 17:26:05 +0400
Subject: [PATCH] Airplane: Remove streams in PoiCompetitorScan


diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java b/src/main/java/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
index ee3b8de9b700202da776c68579532bf11319a001..e436b5277a836d17a1cf79abcb76425d355db7fe 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
@@ -27,13 +27,7 @@ public class PoiCompetitorScan {
                                 world.getPoiManager()
                                     .getType(globalPos.pos())
                                     .ifPresent(
-                                        poiType -> context.<List<LivingEntity>>get(mobs)
-                                                .stream()
-                                                .filter(mob -> mob instanceof Villager && mob != entity)
-                                                .map(villagerx -> (Villager)villagerx)
-                                                .filter(LivingEntity::isAlive)
-                                                .filter(villagerx -> competesForSameJobsite(globalPos, poiType, villagerx))
-                                                .reduce(entity, PoiCompetitorScan::selectWinner)
+                                        poiType -> poiCompete(context.get(mobs), entity, globalPos, poiType) // Leaf - Airplane - Remove streams in PoiCompetitorScan
                                     );
                                 return true;
                             }
@@ -41,6 +35,21 @@ public class PoiCompetitorScan {
         );
     }
 
+    // Leaf start - Airplane - Remove streams in PoiCompetitorScan
+    private static void poiCompete(
+            List<LivingEntity> mobs, Villager entity, GlobalPos globalPos,
+            Holder<net.minecraft.world.entity.ai.village.poi.PoiType> poiType
+    ) {
+        for (LivingEntity mob : mobs) {
+            if (mob instanceof Villager villagerx && mob != entity) {
+                if (villagerx.isAlive() && competesForSameJobsite(globalPos, poiType, villagerx)) {
+                    entity = selectWinner(entity, villagerx);
+                }
+            }
+        }
+    }
+    // Leaf end - Airplane - Remove streams in PoiCompetitorScan
+
     private static Villager selectWinner(Villager first, Villager second) {
         Villager villager;
         Villager villager2;
