From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 12 Jun 2023 17:57:03 +0400
Subject: [PATCH] Last-Part-Of-Server-Crashes-Fix


diff --git a/src/main/java/io/papermc/paper/util/TickThread.java b/src/main/java/io/papermc/paper/util/TickThread.java
index 6a2873e7a184223cbede0e7d21ad54e8308bc69a..78c98e33dc58ee14f9d5ca170171974084b12758 100644
--- a/src/main/java/io/papermc/paper/util/TickThread.java
+++ b/src/main/java/io/papermc/paper/util/TickThread.java
@@ -136,10 +136,24 @@ public class TickThread extends Thread {
     }
 
     public static boolean isTickThreadFor(final ServerLevel world, final int chunkX, final int chunkZ) {
-        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
+        ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
             TickRegionScheduler.getCurrentRegion();
         if (region == null) {
-            return isShutdownThread();
+            try{
+                if (!Bukkit.isOwnedByCurrentRegion(world.getWorld(), chunkX, chunkZ)){
+                    region = world.regioniser.getRegionAtSynchronised(chunkX, chunkZ);
+                    if (region == null){
+                        MinecraftServer.LOGGER.warn("Region at position [" + chunkX + ", " + chunkZ + "] is null!");
+                        return isShutdownThread();
+                    }
+                }
+                else{
+                    return false;
+                }
+            } catch (Exception e) {
+                MinecraftServer.LOGGER.warn("Region at position [" + chunkX + ", " + chunkZ + "] is null!");
+                return isShutdownThread();
+            }
         }
         return world.regioniser.getRegionAtUnsynchronised(chunkX, chunkZ) == region;
     }
