From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Thu, 4 Jan 2024 16:39:26 +0400
Subject: [PATCH] Change-reload-config-work


diff --git a/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java b/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java
index ad854d09999ce2bee9e48976eb79ef02426947a4..b9406cfec2606fabe7fcd22591d19f1ea2f04834 100644
--- a/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java
+++ b/src/main/java/net/edenor/foldenor/commands/FoldenorReloadConfigCommand.java
@@ -35,7 +35,7 @@ public class FoldenorReloadConfigCommand extends Command {
             return false;
         }
         try {
-            FoldenorConfig.reload((java.io.File) DedicatedServer.getServer().options.valueOf("foldenor-settings"));
+            FoldenorConfig.readConfig();
             Command.broadcastCommandMessage(sender, text("Foldenor config reload complete.", GREEN));
             return true;
         } catch (Exception e) {
diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 79073eca74752d550a21f357b5370a76a1457863..7c50e7ab39070b72508a467fe0a93575fc09d58c 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -89,7 +89,7 @@ public class FoldenorConfig {
         Bukkit.getLogger().log(level, s);
     }
 
-    static void readConfig() {
+    public static void readConfig() {
         networkSettings();
 
         optimizationSettings();
