From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 10 Jul 2023 17:41:17 +0400
Subject: [PATCH] Redo-read-config-method


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 565e86cc94ccb5cd333ae64bbb0b1220de96669f..b4c760a05e204a7a355f37c31ef2929d56bc2388 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -59,7 +59,7 @@ public class FoldenorConfig {
         version = getInt("config-version", 1);
         set("config-version", 1);
 
-        readConfig(FoldenorConfig.class, null, setup);
+        readConfig();
     }
 
     protected static void log(String s) {
@@ -72,22 +72,12 @@ public class FoldenorConfig {
         Bukkit.getLogger().log(level, s);
     }
 
-    static void readConfig(Class<?> clazz, Object instance, boolean setup) {
-        for (Method method : clazz.getDeclaredMethods()) {
-            if (Modifier.isPrivate(method.getModifiers())) {
-                if (method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE) {
-                    if (method.getName().endsWith("Setup") && !setup) continue;
-                    try {
-                        method.setAccessible(true);
-                        method.invoke(instance);
-                    } catch (InvocationTargetException ex) {
-                        throw Throwables.propagate(ex.getCause());
-                    } catch (Exception ex) {
-                        Bukkit.getLogger().log(Level.SEVERE, "Error invoking " + method, ex);
-                    }
-                }
-            }
-        }
+    static void readConfig() {
+        networkSettings();
+
+        optimizationSettings();
+
+        otherSettings();
 
         try {
             config.save(CONFIG_FILE);
