From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sat, 4 Jan 2025 14:49:52 +0400
Subject: [PATCH] Pufferfish: Remove lambda from ticking guard


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 8f743959197fab8a516105b4ee11237bca3a5146..36ef59f82f3e8bda974962fd2cddbc8b4b8c35f3 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -834,7 +834,21 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                             }
 
                             gameprofilerfiller.push("tick");
-                            this.guardEntityTick(this::tickNonPassenger, entity);
+                            // this.guardEntityTick(this::tickNonPassenger, entity);
+                            // Pufferfish start - copied from this.guardEntityTick
+                            try {
+                                this.tickNonPassenger(entity); // Pufferfish - changed
+                            } catch (Throwable throwable) {
+                                if (throwable instanceof ThreadDeath) throw throwable; // Paper
+                                // Paper start - Prevent tile entity and entity crashes
+                                final String msg = String.format("Entity threw exception at %s:%s,%s,%s", entity.level().getWorld().getName(), entity.getX(), entity.getY(), entity.getZ());
+                                MinecraftServer.LOGGER.error(msg, throwable);
+                                getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, throwable)));
+                                entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
+                                // Paper end
+                            }
+                            this.moonrise$midTickTasks(); // Paper - rewrite chunk system
+                            // Pufferfish end
                             gameprofilerfiller.pop();
                         }
                     }
