From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 20 Jun 2023 15:27:54 +0400
Subject: [PATCH] Attemp-to-fix-returning-null-world-data


diff --git a/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java b/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
index ee9f5e1f3387998cddbeb1dc6dc6e2b1ea7cd670..30908ec1190b82d9b31a3e7e6e1289b046e794fc 100644
--- a/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -107,6 +107,8 @@ public final class TickRegionScheduler {
         if (!(currThread instanceof TickThreadRunner tickThreadRunner)) {
             return RegionShutdownThread.getWorldData();
         }
+        if (((TickThreadRunner) currThread).currentTickingWorldRegionizedData != null)
+            return ((TickThreadRunner) currThread).currentTickingWorldRegionizedData;
         return tickThreadRunner.currentTickingWorldRegionizedData;
     }
 
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 8cf4428d8140fd03f0ad8fa1d148a04d8caefda8..a3c4029fccc853b17b3a43fda33a034b8195bac5 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -273,7 +273,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public final java.util.concurrent.ConcurrentHashMap.KeySetView<net.minecraft.server.level.ChunkHolder, Boolean> needsChangeBroadcasting = java.util.concurrent.ConcurrentHashMap.newKeySet();
 
     public io.papermc.paper.threadedregions.RegionizedWorldData getCurrentWorldData() {
-        final io.papermc.paper.threadedregions.RegionizedWorldData ret = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        io.papermc.paper.threadedregions.RegionizedWorldData ret = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
         if (ret == null) {
             return ret;
         }
