From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 7 Jan 2025 14:23:18 +0400
Subject: [PATCH] SparklyPaper: Allow throttling hopper checks if the target
 container is full


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 2c608fc97e62a5c03b49ae177320a8d725f3c081..7e57834021112801e242f76bd56707fced0a2fdd 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -55,6 +55,9 @@ public class FoldenorConfig {
 
     public static boolean skipAIForNonAwareMob = true;
 
+    public static boolean throttleHopperWhenFullEnabled = false;
+    public static int throttleHopperWhenFullSkipTicks = 0;
+
     public static void init(File configFile) {
         init(configFile, true);
     }
@@ -160,6 +163,8 @@ public class FoldenorConfig {
         skipMapItemUpdatesIfNoBukkitRender = getBoolean("optimizations.skip_map_item_updates_if_no_bukkit_render", skipMapItemUpdatesIfNoBukkitRender);
         entityActivationCheckFrequency = getInt("optimizations.entity-activation-check-frequency", 20);
         skipAIForNonAwareMob = getBoolean("optimizations.skip-ai-for-non-aware-mob", skipAIForNonAwareMob);
+        throttleHopperWhenFullEnabled = getBoolean("optimizations.throttle-hopper-when-full.enabled", throttleHopperWhenFullEnabled);
+        throttleHopperWhenFullSkipTicks = getInt("optimizations.throttle-hopper-when-full.skip-ticks", throttleHopperWhenFullSkipTicks);
     }
 
     private static void readMiscSettings() {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index bcd5b6948aee2064b1f698bd43ae6485251049ac..976fdee175a27452d9d9e8253c1ade693411aeaa 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -4,6 +4,8 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.HolderLookup;
@@ -468,6 +470,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             Direction enumdirection = blockEntity.facing.getOpposite();
 
             if (HopperBlockEntity.isFullContainer(iinventory, enumdirection)) {
+                // Leaf start - Throttle hopper when full
+                if (FoldenorConfig.throttleHopperWhenFullEnabled && FoldenorConfig.throttleHopperWhenFullSkipTicks > 0) {
+                    blockEntity.setCooldown(FoldenorConfig.throttleHopperWhenFullSkipTicks);
+                }
+                // Leaf end - Throttle hopper when full
                 return false;
             } else {
                 // Paper start - Perf: Optimize Hoppers
