From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 7 Jan 2025 14:42:49 +0400
Subject: [PATCH] Leaf: Nitori: Async playerdata Save


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index 6096916f66cf4574290ded96c99e07669028a92c..3ef9c6ea92684ae5b708817b2af6ed3be1de2ba8 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -62,6 +62,8 @@ public class FoldenorConfig {
     public static boolean lagCompensationEnableForWater = false;
     public static boolean lagCompensationEnableForLava = false;
 
+    public static boolean asyncPlayerDataSaveEnabled = false;
+
     public static void init(File configFile) {
         init(configFile, true);
     }
@@ -177,6 +179,7 @@ public class FoldenorConfig {
     private static void readMiscSettings() {
         useVirtualThreadForAsyncScheduler = getBoolean("optimizations.use-virtual-thread-for-async-scheduler", useVirtualThreadForAsyncScheduler,
             "Use the new Virtual Thread introduced in JDK 21 for CraftAsyncScheduler.");
+        asyncPlayerDataSaveEnabled = getBoolean("misc.async-playerdata-save.enabled", asyncPlayerDataSaveEnabled);
     }
 
     public static boolean pcaSyncProtocolEnabled = false;
diff --git a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
index cdca5ae69991cc068bfbc0686b5defb3604a5440..5ab705731209fd4d4c20cd342ee7bf1bf26f3b0c 100644
--- a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -605,7 +605,11 @@ public class LevelStorageSource {
             CompoundTag nbttagcompound2 = new CompoundTag();
 
             nbttagcompound2.put("Data", nbttagcompound1);
-            this.saveLevelData(nbttagcompound2);
+
+            // Leaf start - Nitori - Async playerdata save
+            Runnable runnable = () -> this.saveLevelData(nbttagcompound2);
+            org.dreeam.leaf.async.AsyncPlayerDataSaving.saveAsync(runnable);
+            // Leaf end - Nitori - Async playerdata save
         }
 
         private void saveLevelData(CompoundTag nbt) {
@@ -702,7 +706,11 @@ public class LevelStorageSource {
             CompoundTag nbttagcompound = LevelStorageSource.readLevelDataTagRaw(this.levelDirectory.dataFile());
 
             nbtProcessor.accept(nbttagcompound.getCompound("Data"));
-            this.saveLevelData(nbttagcompound);
+
+            // Leaf start - Nitori - Async playerdata save
+            Runnable runnable = () -> this.saveLevelData(nbttagcompound);
+            org.dreeam.leaf.async.AsyncPlayerDataSaving.saveAsync(runnable);
+            // Leaf end - Nitori - Async playerdata save
         }
 
         public long makeWorldBackup() throws IOException {
diff --git a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
index b148cf247acdd36f856d0495cde4cc5ad32b5a2f..e825d9e573a38531f5a3b3f9cdccc24570953015 100644
--- a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -36,6 +36,13 @@ public class PlayerDataStorage {
 
     public void save(Player player) {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
+
+        // Leaf start - Nitori - Async playerdata save
+        Runnable runnable = () -> save0(player);
+        org.dreeam.leaf.async.AsyncPlayerDataSaving.saveAsync(runnable);
+    }
+    private void save0(Player player) {
+        // Leaf end - Nitori - Async playerdata save
         try {
             CompoundTag nbttagcompound = player.saveWithoutId(new CompoundTag());
             Path path = this.playerDir.toPath();
diff --git a/src/main/java/org/dreeam/leaf/async/AsyncPlayerDataSaving.java b/src/main/java/org/dreeam/leaf/async/AsyncPlayerDataSaving.java
new file mode 100644
index 0000000000000000000000000000000000000000..65be3ae7b162b201b3f4bce8cfb67f60089033e7
--- /dev/null
+++ b/src/main/java/org/dreeam/leaf/async/AsyncPlayerDataSaving.java
@@ -0,0 +1,23 @@
+package org.dreeam.leaf.async;
+
+import net.edenor.foldenor.config.FoldenorConfig;
+import net.minecraft.Util;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+public class AsyncPlayerDataSaving {
+
+    private AsyncPlayerDataSaving() {
+    }
+
+    public static void saveAsync(Runnable runnable) {
+        if (!FoldenorConfig.asyncPlayerDataSaveEnabled) {
+            runnable.run();
+            return;
+        }
+
+        ExecutorService ioExecutor = Util.backgroundExecutor().service();
+        CompletableFuture.runAsync(runnable, ioExecutor);
+    }
+}
