From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sat, 15 Jul 2023 01:51:07 +0400
Subject: [PATCH] Remove-some-throws


diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index 49c9088c365fdc4342b82de7f1ce2cc0b93e7ea4..bfdded45dde8dd596055c5015a7b952606f04cad 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -811,40 +811,40 @@ public class ServerChunkCache extends ChunkSource {
         // Folia start - region threading
         @Override
         public void tell(Runnable runnable) {
-            if (true) {
+            /*if (true) {
                 throw new UnsupportedOperationException();
-            }
+            }*/
             super.tell(runnable);
         }
 
         @Override
         public void executeBlocking(Runnable runnable) {
-            if (true) {
+            /*if (true) {
                 throw new UnsupportedOperationException();
-            }
+            }*/
             super.executeBlocking(runnable);
         }
 
         @Override
         public void execute(Runnable runnable) {
-            if (true) {
+            /*if (true) {
                 throw new UnsupportedOperationException();
-            }
+            }*/
             super.execute(runnable);
         }
 
         @Override
         public void executeIfPossible(Runnable runnable) {
-            if (true) {
+            /*if (true) {
                 throw new UnsupportedOperationException();
-            }
+            }*/
             super.executeIfPossible(runnable);
         }
         // Folia end - region threading
 
         @Override
         protected void doRunTask(Runnable task) {
-            if (true) throw new UnsupportedOperationException(); // Folia - region threading
+            //if (true) throw new UnsupportedOperationException(); // Folia - region threading
             ServerChunkCache.this.level.getProfiler().incrementCounter("runTask");
             super.doRunTask(task);
         }
