From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Mon, 14 Jul 2025 15:10:07 +0400
Subject: [PATCH] Let Me Despawn


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index a13c27ed857bd0de290acaae90b4f236b0ae9c54..e5a73fe81910ea8ad7d19537f9d339faf2a014ed 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -720,7 +720,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
             // Let Me Despawn start
-            if (FoldenorUtils.pickedItems)
+            if (FoldenorConfig.lmdEnabled && FoldenorUtils.pickedItems)
                 FoldenorUtils.dropEquipmentOnDiscard(this);
             // Let Me Despawn end
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
@@ -738,7 +738,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
                     // Let Me Despawn start
-                    if (FoldenorUtils.pickedItems)
+                    if (FoldenorConfig.lmdEnabled && FoldenorUtils.pickedItems)
                         FoldenorUtils.dropEquipmentOnDiscard(this);
                     // Let Me Despawn end
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
@@ -747,7 +747,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
                 if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
                     if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
                         // Let Me Despawn start
-                        if (FoldenorUtils.pickedItems)
+                        if (FoldenorConfig.lmdEnabled && FoldenorUtils.pickedItems)
                             FoldenorUtils.dropEquipmentOnDiscard(this);
                         // Let Me Despawn end
                         this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
