--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -40,6 +_,12 @@
         this.data = new SimpleBitStorage(heightBits, 256);
     }
 
+    // Leaf start - Lithium - combined heightmap update
+    public final Predicate<BlockState> isOpaque() {
+        return this.isOpaque;
+    }
+    // Leaf end - Lithium - combined heightmap update
+
     public static void primeHeightmaps(final ChunkAccess chunk, final Set<Heightmap.Types> types) {
         if (!types.isEmpty()) {
             int size = types.size();
@@ -119,7 +_,7 @@
         return this.data.get(index) + this.chunk.getMinY();
     }
 
-    private void setHeight(final int x, final int z, final int height) {
+    public void setHeight(final int x, final int z, final int height) { // Leaf - Lithium - combined heightmap update - private -> public
         this.data.set(getIndex(x, z), height - this.chunk.getMinY());
     }
 
