--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -592,7 +_,10 @@
             CompoundTag dataTag = levelData.createTag(singleplayerUUID);
             CompoundTag root = new CompoundTag();
             root.put("Data", dataTag);
-            this.saveLevelData(root);
+            // Leaf start - Nitori - Async playerdata save
+            Runnable runnable = () -> this.saveLevelData(root);
+            org.dreeam.leaf.async.AsyncPlayerDataSaving.saveAsync(runnable);
+            // Leaf end - Nitori - Async playerdata save
         }
 
         public void saveLevelData(final Dynamic<?> tag) {
@@ -690,7 +_,10 @@
             this.checkLock();
             CompoundTag root = LevelStorageSource.readLevelDataTagRaw(this.levelDirectory.dataFile());
             updater.accept(root.getCompoundOrEmpty("Data"));
-            this.saveLevelData(root);
+            // Leaf start - Nitori - Async playerdata save
+            Runnable runnable = () -> this.saveLevelData(root);
+            org.dreeam.leaf.async.AsyncPlayerDataSaving.saveAsync(runnable);
+            // Leaf end - Nitori - Async playerdata save
         }
 
         public long makeWorldBackup() throws IOException {
