--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -32,6 +_,13 @@
 
     public void save(final Player player) {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
+        // Leaf start - Nitori - Async playerdata save
+        Runnable runnable = () -> save0(player);
+        org.dreeam.leaf.async.AsyncPlayerDataSaving.saveAsync(runnable);
+    }
+
+    private void save0(Player player) {
+        // Leaf end - Nitori - Async playerdata save
         try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(player.problemPath(), LOGGER)) {
             TagValueOutput output = TagValueOutput.createWithContext(reporter, player.registryAccess());
             player.saveWithoutId(output);
