--- a/net/minecraft/world/level/block/state/pattern/BlockPattern.java
+++ b/net/minecraft/world/level/block/state/pattern/BlockPattern.java
@@ -78,8 +_,8 @@
         int dist = Math.max(Math.max(this.width, this.height), this.depth);
 
         for (BlockPos testPos : BlockPos.betweenClosed(origin, origin.offset(dist - 1, dist - 1, dist - 1))) {
-            for (Direction forwards : Direction.values()) {
-                for (Direction up : Direction.values()) {
+            for (Direction forwards : Direction.VALUES) {
+                for (Direction up : Direction.VALUES) {
                     if (up != forwards && up != forwards.getOpposite()) {
                         BlockPattern.BlockPatternMatch match = this.matches(testPos, forwards, up, cache);
                         if (match != null) {
